{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "CMq8ReGENcIw",
    "outputId": "50d08e3e-0aa6-4d3b-8ed2-daba314b6cb9"
   },
   "outputs": [],
   "source": [
    "# !pip install episcanpy\n",
    "# !pip install scanpy\n",
    "# !pip install anndata\n",
    "# !pip install desc\n",
    "# !pip install display\n",
    "# !pip install --pre deepchem\n",
    "# !pip install nfp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "J5gtEHG-HLb8",
    "outputId": "0c6a7ea7-d156-4aa2-bc95-6537a2a369a7"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.9.1 anndata==0.8.0 umap==0.5.3 numpy==1.21.6 scipy==1.4.1 pandas==1.3.5 scikit-learn==1.0.2 statsmodels==0.10.2 python-igraph==0.9.10 louvain==0.7.1 pynndescent==0.5.6\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import gzip\n",
    "import csv\n",
    "import os\n",
    "import desc as desc\n",
    "import matplotlib.pyplot as plt\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import RDKFingerprint\n",
    "from sklearn.metrics import classification_report, plot_confusion_matrix, confusion_matrix\n",
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "VtvbvkY1NpdK",
    "outputId": "d504264d-d67e-4df1-fdb3-324999c341c9"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mounted at /content/drive\n",
      "'Dataset for GNN'  'Drug Synergy Prediction'\n"
     ]
    }
   ],
   "source": [
    "from google.colab import drive\n",
    "drive.mount('/content/drive')\n",
    "# drive.mount('/content/drive/My Drive/Projects and research stuffs')\n",
    "!ls \"/content/drive/My Drive/Projects and research stuffs/Drug Synergy\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "4UORX-s8hf3l"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "NMcUfNIxPb0V"
   },
   "source": [
    "## What is a Fingerprint?\n",
    "\n",
    "Deep learning models almost always take arrays of numbers as their inputs. If we want to process molecules with them, we somehow need to represent each molecule as one or more arrays of numbers.\n",
    "\n",
    "Many (but not all) types of models require their inputs to have a fixed size. This can be a challenge for molecules, since different molecules have different numbers of atoms. If we want to use these types of models, we somehow need to represent variable sized molecules with fixed sized arrays.\n",
    "\n",
    "Fingerprints are designed to address these problems. A fingerprint is a fixed length array, where different elements indicate the presence of different features in the molecule. If two molecules have similar fingerprints, that indicates they contain many of the same features, and therefore will likely have similar chemistry.\n",
    "\n",
    "DeepChem supports a particular type of fingerprint called an \"Extended Connectivity Fingerprint\", or \"ECFP\" for short. They also are sometimes called \"circular fingerprints\". The ECFP algorithm begins by classifying atoms based only on their direct properties and bonds. Each unique pattern is a feature. For example, \"carbon atom bonded to two hydrogens and two heavy atoms\" would be a feature, and a particular element of the fingerprint is set to 1 for any molecule that contains that feature. It then iteratively identifies new features by looking at larger circular neighborhoods. One specific feature bonded to two other specific features becomes a higher level feature, and the corresponding element is set for any molecule that contains it. This continues for a fixed number of iterations, most often two."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "k5ToLzrOknKj"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 35
    },
    "id": "1IVZ8PZEk8U8",
    "outputId": "139615ff-be1f-4329-b5f2-3aa2bdadd2c4"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "type": "string"
      },
      "text/plain": [
       "'2.6.1'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import deepchem as dc\n",
    "dc.__version__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "7Rv27Z-ucIS4"
   },
   "source": [
    "#### Creating Circular Fingerprints ( Neural based fingerprint creation - GNN )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 223
    },
    "id": "V3ZgKiyhcD31",
    "outputId": "58a2f247-f40d-4b88-c1c2-0dd9d79f2cc7"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(23052, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-d80495e9-4150-43b1-851e-5eb7ec8e8590\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>drug_a_name</th>\n",
       "      <th>drug_b_name</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>synergy</th>\n",
       "      <th>fold</th>\n",
       "      <th>drug_a_structure</th>\n",
       "      <th>drug_b_structure</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5-FU_ABT-888_A2058</td>\n",
       "      <td>5-FU</td>\n",
       "      <td>ABT-888</td>\n",
       "      <td>A2058</td>\n",
       "      <td>7.693530</td>\n",
       "      <td>2</td>\n",
       "      <td>O=c1[nH]cc(F)c(=O)[nH]1</td>\n",
       "      <td>CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5-FU_ABT-888_A2780</td>\n",
       "      <td>5-FU</td>\n",
       "      <td>ABT-888</td>\n",
       "      <td>A2780</td>\n",
       "      <td>7.778053</td>\n",
       "      <td>2</td>\n",
       "      <td>O=c1[nH]cc(F)c(=O)[nH]1</td>\n",
       "      <td>CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5-FU_ABT-888_A375</td>\n",
       "      <td>5-FU</td>\n",
       "      <td>ABT-888</td>\n",
       "      <td>A375</td>\n",
       "      <td>-1.198505</td>\n",
       "      <td>2</td>\n",
       "      <td>O=c1[nH]cc(F)c(=O)[nH]1</td>\n",
       "      <td>CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5-FU_ABT-888_A427</td>\n",
       "      <td>5-FU</td>\n",
       "      <td>ABT-888</td>\n",
       "      <td>A427</td>\n",
       "      <td>2.595684</td>\n",
       "      <td>2</td>\n",
       "      <td>O=c1[nH]cc(F)c(=O)[nH]1</td>\n",
       "      <td>CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5-FU_ABT-888_CAOV3</td>\n",
       "      <td>5-FU</td>\n",
       "      <td>ABT-888</td>\n",
       "      <td>CAOV3</td>\n",
       "      <td>-5.139971</td>\n",
       "      <td>2</td>\n",
       "      <td>O=c1[nH]cc(F)c(=O)[nH]1</td>\n",
       "      <td>CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d80495e9-4150-43b1-851e-5eb7ec8e8590')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-d80495e9-4150-43b1-851e-5eb7ec8e8590 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-d80495e9-4150-43b1-851e-5eb7ec8e8590');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "           Unnamed: 0 drug_a_name drug_b_name cell_line   synergy  fold  \\\n",
       "0  5-FU_ABT-888_A2058        5-FU     ABT-888     A2058  7.693530     2   \n",
       "1  5-FU_ABT-888_A2780        5-FU     ABT-888     A2780  7.778053     2   \n",
       "2   5-FU_ABT-888_A375        5-FU     ABT-888      A375 -1.198505     2   \n",
       "3   5-FU_ABT-888_A427        5-FU     ABT-888      A427  2.595684     2   \n",
       "4  5-FU_ABT-888_CAOV3        5-FU     ABT-888     CAOV3 -5.139971     2   \n",
       "\n",
       "          drug_a_structure                    drug_b_structure  \n",
       "0  O=c1[nH]cc(F)c(=O)[nH]1  CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1  \n",
       "1  O=c1[nH]cc(F)c(=O)[nH]1  CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1  \n",
       "2  O=c1[nH]cc(F)c(=O)[nH]1  CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1  \n",
       "3  O=c1[nH]cc(F)c(=O)[nH]1  CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1  \n",
       "4  O=c1[nH]cc(F)c(=O)[nH]1  CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data_path = \"/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN\"\n",
    "main_data = pd.read_csv(train_data_path+'/final_data.csv')\n",
    "print(main_data.shape)\n",
    "main_data.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 206
    },
    "id": "FNRTYmoPcD6D",
    "outputId": "21f3958b-f297-4086-f152-14b2eb824726"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-098024c5-254d-487d-a237-dbd0cb6f58b2\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_a_structure</th>\n",
       "      <th>drug_b_structure</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>synergy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CN1C(=O)C=CC2(C)C3CCC4(C)C(NC(=O)OCC(F)(F)F)CC...</td>\n",
       "      <td>N.N.O=C(O)C1(C(=O)O)CCC1.[Pt]</td>\n",
       "      <td>A2780</td>\n",
       "      <td>0.121266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1</td>\n",
       "      <td>COC1=C2CC(C)CC(OC)C(O)C(C)C=C(C)C(OC(N)=O)C(OC...</td>\n",
       "      <td>SKOV3</td>\n",
       "      <td>-0.694117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CN1C(=O)C=CC2(C)C3CCC4(C)C(NC(=O)OCC(F)(F)F)CC...</td>\n",
       "      <td>CCc1c2c(nc3ccc(O)cc13)-c1cc3c(c(=O)n1C2)COC(=O...</td>\n",
       "      <td>UWB1289</td>\n",
       "      <td>-0.088576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COc1cccc2c1C(=O)c1c(O)c3c(c(O)c1C2=O)CC(O)(C(=...</td>\n",
       "      <td>COC1CC2CCC(C)C(O)(O2)C(=O)C(=O)N2CCCCC2C(=O)OC...</td>\n",
       "      <td>NCIH460</td>\n",
       "      <td>0.277679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCC1(O)C(=O)OCc2c1cc1n(c2=O)Cc2cc3c(CN(C)C)c(O...</td>\n",
       "      <td>Cn1c(=O)n(-c2ccc(C(C)(C)C#N)cc2)c2c3cc(-c4cnc5...</td>\n",
       "      <td>PA1</td>\n",
       "      <td>-0.649434</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-098024c5-254d-487d-a237-dbd0cb6f58b2')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-098024c5-254d-487d-a237-dbd0cb6f58b2 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-098024c5-254d-487d-a237-dbd0cb6f58b2');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "                                    drug_a_structure  \\\n",
       "0  CN1C(=O)C=CC2(C)C3CCC4(C)C(NC(=O)OCC(F)(F)F)CC...   \n",
       "1                 CC1(c2nc3c(C(N)=O)cccc3[nH]2)CCCN1   \n",
       "2  CN1C(=O)C=CC2(C)C3CCC4(C)C(NC(=O)OCC(F)(F)F)CC...   \n",
       "3  COc1cccc2c1C(=O)c1c(O)c3c(c(O)c1C2=O)CC(O)(C(=...   \n",
       "4  CCC1(O)C(=O)OCc2c1cc1n(c2=O)Cc2cc3c(CN(C)C)c(O...   \n",
       "\n",
       "                                    drug_b_structure cell_line   synergy  \n",
       "0                      N.N.O=C(O)C1(C(=O)O)CCC1.[Pt]     A2780  0.121266  \n",
       "1  COC1=C2CC(C)CC(OC)C(O)C(C)C=C(C)C(OC(N)=O)C(OC...     SKOV3 -0.694117  \n",
       "2  CCc1c2c(nc3ccc(O)cc13)-c1cc3c(c(=O)n1C2)COC(=O...   UWB1289 -0.088576  \n",
       "3  COC1CC2CCC(C)C(O)(O2)C(=O)C(=O)N2CCCCC2C(=O)OC...   NCIH460  0.277679  \n",
       "4  Cn1c(=O)n(-c2ccc(C(C)(C)C#N)cc2)c2c3cc(-c4cnc5...       PA1 -0.649434  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.preprocessing import StandardScaler\n",
    "data_for_gnn = main_data[['drug_a_structure', 'drug_b_structure', 'cell_line', 'synergy']].copy()\n",
    "data_for_gnn = data_for_gnn.sample(frac=1).reset_index().drop(['index'], 1)\n",
    "\n",
    "scaler = StandardScaler().fit(data_for_gnn['synergy'].values.reshape(-1,1))\n",
    "scaled_ops = scaler.transform(data_for_gnn['synergy'].values.reshape(-1,1))\n",
    "data_for_gnn['synergy'] = scaled_ops\n",
    "data_for_gnn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "n8pN4m7q4wQN",
    "outputId": "d9722a48-502a-4f95-cc0e-d39dbad5bb95"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23052, 4)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_for_gnn.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "id": "SpvHErjyXHcN"
   },
   "outputs": [],
   "source": [
    "#### Need to remove all drug smiles with '.' in them - this '.' creates separate isolated and disconnected graphs and it gets hard to calculate adjacency matrix and update them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "uTm3jzwwcD8l"
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "from rdkit import Chem\n",
    "\n",
    "import torch\n",
    "\n",
    "\n",
    "def create_atoms(mol, atom_dict):\n",
    "    \"\"\"Transform the atom types in a molecule (e.g., H, C, and O)\n",
    "    into the indices (e.g., H=0, C=1, and O=2).\n",
    "    Note that each atom index considers the aromaticity.\n",
    "    \"\"\"\n",
    "    atoms = [a.GetSymbol() for a in mol.GetAtoms()]\n",
    "    for a in mol.GetAromaticAtoms():\n",
    "        i = a.GetIdx()\n",
    "        atoms[i] = (atoms[i], 'aromatic')\n",
    "    atoms = [atom_dict[a] for a in atoms]\n",
    "    return np.array(atoms)\n",
    "\n",
    "\n",
    "def create_ijbonddict(mol, bond_dict):\n",
    "    \"\"\"Create a dictionary, in which each key is a node ID\n",
    "    and each value is the tuples of its neighboring node\n",
    "    and chemical bond (e.g., single and double) IDs.\n",
    "    \"\"\"\n",
    "    i_jbond_dict = defaultdict(lambda: [])\n",
    "    for b in mol.GetBonds():\n",
    "        i, j = b.GetBeginAtomIdx(), b.GetEndAtomIdx()\n",
    "        bond = bond_dict[str(b.GetBondType())]\n",
    "        i_jbond_dict[i].append((j, bond))\n",
    "        i_jbond_dict[j].append((i, bond))\n",
    "    return i_jbond_dict\n",
    "\n",
    "\n",
    "def extract_fingerprints(radius, atoms, i_jbond_dict,\n",
    "                         fingerprint_dict, edge_dict):\n",
    "    \"\"\"Extract the fingerprints from a molecular graph\n",
    "    based on Weisfeiler-Lehman algorithm.\n",
    "    \"\"\"\n",
    "\n",
    "    if (len(atoms) == 1) or (radius == 0):\n",
    "        nodes = [fingerprint_dict[a] for a in atoms]\n",
    "\n",
    "    else:\n",
    "        nodes = atoms\n",
    "        i_jedge_dict = i_jbond_dict\n",
    "\n",
    "        for _ in range(radius):\n",
    "\n",
    "            \"\"\"Update each node ID considering its neighboring nodes and edges.\n",
    "            The updated node IDs are the fingerprint IDs.\n",
    "            \"\"\"\n",
    "            nodes_ = []\n",
    "            for i, j_edge in i_jedge_dict.items():\n",
    "                neighbors = [(nodes[j], edge) for j, edge in j_edge]\n",
    "                fingerprint = (nodes[i], tuple(sorted(neighbors)))\n",
    "                nodes_.append(fingerprint_dict[fingerprint])\n",
    "\n",
    "            \"\"\"Also update each edge ID considering\n",
    "            its two nodes on both sides.\n",
    "            \"\"\"\n",
    "            i_jedge_dict_ = defaultdict(lambda: [])\n",
    "            for i, j_edge in i_jedge_dict.items():\n",
    "                for j, edge in j_edge:\n",
    "                    both_side = tuple(sorted((nodes[i], nodes[j])))\n",
    "                    edge = edge_dict[(both_side, edge)]\n",
    "                    i_jedge_dict_[i].append((j, edge))\n",
    "\n",
    "            nodes = nodes_\n",
    "            i_jedge_dict = i_jedge_dict_\n",
    "\n",
    "    return np.array(nodes)\n",
    "\n",
    "\n",
    "def split_dataset(dataset, ratio):\n",
    "    \"\"\"Shuffle and split a dataset.\"\"\"\n",
    "    np.random.seed(1234)  # fix the seed for shuffle.\n",
    "    np.random.shuffle(dataset)\n",
    "    n = int(ratio * len(dataset))\n",
    "    return dataset[:n], dataset[n:]\n",
    "\n",
    "\n",
    "def create_datasets(task, data_for_gnn, radius, device):\n",
    "\n",
    "    dir_dataset =  ''\n",
    "    print(type(data_for_gnn))\n",
    "    print(data_for_gnn.shape)\n",
    "    \"\"\"Initialize x_dict, in which each key is a symbol type\n",
    "    (e.g., atom and chemical bond) and each value is its index.\n",
    "    \"\"\"\n",
    "    atom_dict = defaultdict(lambda: len(atom_dict))\n",
    "    bond_dict = defaultdict(lambda: len(bond_dict))\n",
    "    fingerprint_dict = defaultdict(lambda: len(fingerprint_dict))\n",
    "    edge_dict = defaultdict(lambda: len(edge_dict))\n",
    "   \n",
    "    \n",
    "    non_working_smiles_index = []\n",
    "    def create_dataset(data):\n",
    "\n",
    "#         print(filename)\n",
    "\n",
    "#         \"\"\"Load a dataset.\"\"\"\n",
    "#         with open(dir_dataset + filename, 'r') as f:\n",
    "#             smiles_property = f.readline().strip().split()\n",
    "#             data_original = f.read().strip().split('\\n')\n",
    "\n",
    "#         \"\"\"Exclude the data contains '.' in its smiles.\"\"\"\n",
    "#         data_original = [data for data in data_original\n",
    "#                          if '.' not in data.split()[0]]\n",
    "\n",
    "        dataset = []\n",
    "\n",
    "        for data_index in range(len(data)):\n",
    "#             print(data_index)\n",
    "            try:\n",
    "        #             smiles, property = data.strip().split()\n",
    "                smiles1, smiles2, property = data.iloc[data_index]['drug_a_structure'], data.iloc[data_index]['drug_b_structure'], data.iloc[data_index]['synergy']\n",
    "\n",
    "                \"\"\"Create each data with the above defined functions.\"\"\"\n",
    "                # for drug A:\n",
    "                mol = Chem.AddHs(Chem.MolFromSmiles(smiles1))\n",
    "                atoms = create_atoms(mol, atom_dict)\n",
    "                molecular_size1 = len(atoms)\n",
    "                i_jbond_dict = create_ijbonddict(mol, bond_dict)\n",
    "                fingerprints1 = extract_fingerprints(radius, atoms, i_jbond_dict,\n",
    "                                                    fingerprint_dict, edge_dict)\n",
    "                adjacency1 = Chem.GetAdjacencyMatrix(mol)\n",
    "\n",
    "                # for drug B:\n",
    "                mol = Chem.AddHs(Chem.MolFromSmiles(smiles2))\n",
    "                atoms = create_atoms(mol, atom_dict)\n",
    "                molecular_size2 = len(atoms)\n",
    "                i_jbond_dict = create_ijbonddict(mol, bond_dict)\n",
    "                fingerprints2 = extract_fingerprints(radius, atoms, i_jbond_dict,\n",
    "                                                    fingerprint_dict, edge_dict)\n",
    "                adjacency2 = Chem.GetAdjacencyMatrix(mol)\n",
    "\n",
    "                \"\"\"Transform the above each data of numpy\n",
    "                to pytorch tensor on a device (i.e., CPU or GPU).\n",
    "                \"\"\"\n",
    "                fingerprints1 = torch.LongTensor(fingerprints1).to(device)\n",
    "                adjacency1 = torch.FloatTensor(adjacency1).to(device)\n",
    "                fingerprints2 = torch.LongTensor(fingerprints2).to(device)\n",
    "                adjacency2 = torch.FloatTensor(adjacency2).to(device)\n",
    "\n",
    "                if task == 'regression':\n",
    "                    property = torch.FloatTensor([[float(property)]]).to(device)\n",
    "\n",
    "                dataset.append((fingerprints1, adjacency1, molecular_size1, fingerprints2, adjacency2, molecular_size2, property))\n",
    "            \n",
    "            except:\n",
    "                non_working_smiles_index.append(data_index)\n",
    "            \n",
    "        return dataset\n",
    "\n",
    "    dataset_train = create_dataset(data_for_gnn[:22000])\n",
    "    print('Skipped '+ str(len(non_working_smiles_index)) + ' smiles in training data')\n",
    "    dataset_train, dataset_dev = split_dataset(dataset_train, 0.9)\n",
    "    dataset_test = create_dataset(data_for_gnn[22000:])\n",
    "\n",
    "    N_fingerprints = len(fingerprint_dict)\n",
    "\n",
    "    return dataset_train, dataset_dev, dataset_test, N_fingerprints\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "fvatyGKGcD-8"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "KKkcQc9LcEA-"
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import timeit\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch.optim as optim\n",
    "\n",
    "from sklearn.metrics import roc_auc_score\n",
    "\n",
    "# import preprocess as pp\n",
    "\n",
    "\n",
    "class MolecularGraphNeuralNetwork(nn.Module):\n",
    "    def __init__(self, N_fingerprints, dim, layer_hidden, layer_output):\n",
    "        super(MolecularGraphNeuralNetwork, self).__init__()\n",
    "        self.embed_fingerprint = nn.Embedding(N_fingerprints, dim)\n",
    "        self.W_fingerprint = nn.ModuleList([nn.Linear(dim, dim)\n",
    "                                            for _ in range(layer_hidden)])\n",
    "        self.W_output = nn.ModuleList([nn.Linear(dim+dim, dim+dim)\n",
    "                                       for _ in range(layer_output)])\n",
    "        if task == 'classification':\n",
    "            self.W_property = nn.Linear(dim, 2)\n",
    "        if task == 'regression':\n",
    "            self.W_property = nn.Linear(dim+dim, 1)\n",
    "\n",
    "    def pad(self, matrices, pad_value):\n",
    "        \"\"\"Pad the list of matrices\n",
    "        with a pad_value (e.g., 0) for batch processing.\n",
    "        For example, given a list of matrices [A, B, C],\n",
    "        we obtain a new matrix [A00, 0B0, 00C],\n",
    "        where 0 is the zero (i.e., pad value) matrix.\n",
    "        \"\"\"\n",
    "        shapes = [m.shape for m in matrices]\n",
    "        M, N = sum([s[0] for s in shapes]), sum([s[1] for s in shapes])\n",
    "        zeros = torch.FloatTensor(np.zeros((M, N))).to(device)\n",
    "        pad_matrices = pad_value + zeros\n",
    "        i, j = 0, 0\n",
    "        for k, matrix in enumerate(matrices):\n",
    "            m, n = shapes[k]\n",
    "            pad_matrices[i:i+m, j:j+n] = matrix\n",
    "            i += m\n",
    "            j += n\n",
    "        return pad_matrices\n",
    "\n",
    "    def update(self, matrix, vectors, layer):\n",
    "        hidden_vectors = torch.relu(self.W_fingerprint[layer](vectors))\n",
    "        return hidden_vectors + torch.matmul(matrix, hidden_vectors)\n",
    "\n",
    "    def sum(self, vectors, axis):\n",
    "        sum_vectors = [torch.sum(v, 0) for v in torch.split(vectors, axis)]\n",
    "        return torch.stack(sum_vectors)\n",
    "\n",
    "    def mean(self, vectors, axis):\n",
    "        mean_vectors = [torch.mean(v, 0) for v in torch.split(vectors, axis)]\n",
    "        return torch.stack(mean_vectors)\n",
    "\n",
    "    def gnn(self, inputs):\n",
    "\n",
    "        \"\"\"Cat or pad each input data for batch processing.\"\"\"\n",
    "        fingerprints, adjacencies, molecular_sizes = inputs\n",
    "        fingerprints = torch.cat(fingerprints)\n",
    "        adjacencies = self.pad(adjacencies, 0)\n",
    "\n",
    "        \"\"\"GNN layer (update the fingerprint vectors).\"\"\"\n",
    "        fingerprint_vectors = self.embed_fingerprint(fingerprints)\n",
    "        for l in range(layer_hidden):\n",
    "            hs = self.update(adjacencies, fingerprint_vectors, l)\n",
    "            fingerprint_vectors = F.normalize(hs, 2, 1)  # normalize.\n",
    "\n",
    "        \"\"\"Molecular vector by sum or mean of the fingerprint vectors.\"\"\"\n",
    "        molecular_vectors = self.sum(fingerprint_vectors, molecular_sizes)\n",
    "        # molecular_vectors = self.mean(fingerprint_vectors, molecular_sizes)\n",
    "\n",
    "        return molecular_vectors\n",
    "\n",
    "    def mlp(self, vectors):\n",
    "        \"\"\"Classifier or regressor based on multilayer perceptron.\"\"\"\n",
    "        for l in range(layer_output):\n",
    "#             print(self.W_output[l](vectors))\n",
    "#             print(self.W_output[l](vectors).shape)\n",
    "            vectors = torch.relu(self.W_output[l](vectors))\n",
    "        outputs = self.W_property(vectors)\n",
    "        return outputs\n",
    "\n",
    "    def forward_classifier(self, data_batch, train):\n",
    "\n",
    "        inputs = data_batch[:-1]\n",
    "        correct_labels = torch.cat(data_batch[-1])\n",
    "\n",
    "        if train:\n",
    "            molecular_vectors = self.gnn(inputs)\n",
    "            predicted_scores = self.mlp(molecular_vectors)\n",
    "            loss = F.cross_entropy(predicted_scores, correct_labels)\n",
    "            return loss\n",
    "        else:\n",
    "            with torch.no_grad():\n",
    "                molecular_vectors = self.gnn(inputs)\n",
    "                predicted_scores = self.mlp(molecular_vectors)\n",
    "            predicted_scores = predicted_scores.to('cpu').data.numpy()\n",
    "            predicted_scores = [s[1] for s in predicted_scores]\n",
    "            correct_labels = correct_labels.to('cpu').data.numpy()\n",
    "            return predicted_scores, correct_labels\n",
    "\n",
    "    def forward_regressor(self, data_batch, train):\n",
    "\n",
    "        inputs_d1 = data_batch[:3]\n",
    "        inputs_d2 = data_batch[3:-1]\n",
    "        correct_values = torch.cat(data_batch[-1])\n",
    "\n",
    "        if train:\n",
    "            molecular_vectors_d1 = self.gnn(inputs_d1)\n",
    "            molecular_vectors_d2 = self.gnn(inputs_d2)\n",
    "#             molecular_vectors = np.concatenate([molecular_vectors_d1,molecular_vectors_d2])\n",
    "            molecular_vectors = torch.cat((molecular_vectors_d1, molecular_vectors_d2), 1)\n",
    "            predicted_values = self.mlp(molecular_vectors)\n",
    "#             print(predicted_values.shape, correct_values.shape, molecular_vectors.shape, molecular_vectors_d1.shape)\n",
    "            loss = F.mse_loss(predicted_values, correct_values)\n",
    "            return loss\n",
    "        else:\n",
    "            with torch.no_grad():\n",
    "                molecular_vectors_d1 = self.gnn(inputs_d1)\n",
    "                molecular_vectors_d2 = self.gnn(inputs_d2)\n",
    "                molecular_vectors = torch.cat((molecular_vectors_d1, molecular_vectors_d2), 1)\n",
    "                predicted_values = self.mlp(molecular_vectors)\n",
    "            predicted_values = predicted_values.to('cpu').data.numpy()\n",
    "            correct_values = correct_values.to('cpu').data.numpy()\n",
    "            predicted_values = np.concatenate(predicted_values)\n",
    "            correct_values = np.concatenate(correct_values)\n",
    "            return predicted_values, correct_values\n",
    "\n",
    "\n",
    "class Trainer(object):\n",
    "    def __init__(self, model):\n",
    "        self.model = model\n",
    "        self.optimizer = optim.Adam(self.model.parameters(), lr=lr)\n",
    "\n",
    "    def train(self, dataset):\n",
    "        np.random.shuffle(dataset)\n",
    "        N = len(dataset)\n",
    "        loss_total = 0\n",
    "        for i in range(0, N, batch_train):\n",
    "            data_batch = list(zip(*dataset[i:i+batch_train]))\n",
    "            if task == 'classification':\n",
    "                loss = self.model.forward_classifier(data_batch, train=True)\n",
    "            if task == 'regression':\n",
    "                loss = self.model.forward_regressor(data_batch, train=True)\n",
    "            self.optimizer.zero_grad()\n",
    "            loss.backward()\n",
    "            self.optimizer.step()\n",
    "            loss_total += loss.item()\n",
    "        return loss_total\n",
    "\n",
    "\n",
    "class Tester(object):\n",
    "    def __init__(self, model):\n",
    "        self.model = model\n",
    "\n",
    "    def test_regressor(self, dataset):\n",
    "        N = len(dataset)\n",
    "        SAE = 0  # sum absolute error.\n",
    "        for i in range(0, N, batch_test):\n",
    "            data_batch = list(zip(*dataset[i:i+batch_test]))\n",
    "            predicted_values, correct_values = self.model.forward_regressor(\n",
    "                                               data_batch, train=False)\n",
    "            SAE += sum(np.abs(predicted_values-correct_values))\n",
    "        MAE = SAE / N  # mean absolute error.\n",
    "        return MAE\n",
    "\n",
    "    def save_result(self, result, filename):\n",
    "        with open(filename, 'a') as f:\n",
    "            f.write(result + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "Wo4IN_ubcEC_",
    "outputId": "403bf613-96cd-4cef-f11d-72d7abf7f751"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The code uses a GPU!\n",
      "Preprocessing the  dataset.\n",
      "Just a moment......\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "(23052, 3)\n",
      "Skipped 1639 smiles in training data\n",
      "----------------------------------------------------------------------------------------------------\n",
      "The preprocess has finished!\n",
      "# of training data samples: 18324\n",
      "# of development data samples: 2037\n",
      "# of test data samples: 975\n",
      "----------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Run the main code function:\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    task = 'regression'\n",
    "    dataset = ''\n",
    "    radius = 3\n",
    "    dim = 50\n",
    "    layer_hidden=6\n",
    "    layer_output=6\n",
    "\n",
    "    batch_train=32\n",
    "    batch_test=32\n",
    "    lr=1e-4\n",
    "    lr_decay=0.99\n",
    "    decay_interval=10\n",
    "    iteration=5\n",
    "\n",
    "#     (task, dataset, radius, dim, layer_hidden, layer_output,\n",
    "#      batch_train, batch_test, lr, lr_decay, decay_interval, iteration,\n",
    "#      setting) = sys.argv[1:]\n",
    "    (radius, dim, layer_hidden, layer_output,\n",
    "     batch_train, batch_test, decay_interval,\n",
    "     iteration) = map(int, [radius, dim, layer_hidden, layer_output,\n",
    "                            batch_train, batch_test,\n",
    "                            decay_interval, iteration])\n",
    "    lr, lr_decay = map(float, [lr, lr_decay])\n",
    "#     device = torch.device('cuda')\n",
    "    # device = torch.device('cpu')\n",
    "    \n",
    "    if torch.cuda.is_available():\n",
    "        device = torch.device('cuda')\n",
    "        print('The code uses a GPU!')\n",
    "    else:\n",
    "        device = torch.device('cpu')\n",
    "        print('The code uses a CPU...')\n",
    "#     print('-'*100)\n",
    "\n",
    "    print('Preprocessing the', dataset, 'dataset.')\n",
    "    print('Just a moment......')\n",
    "    (dataset_train, dataset_dev, dataset_test,\n",
    "     N_fingerprints) = create_datasets(task, data_for_gnn, radius, device)\n",
    "    print('-'*100)\n",
    "\n",
    "    print('The preprocess has finished!')\n",
    "    print('# of training data samples:', len(dataset_train))\n",
    "    print('# of development data samples:', len(dataset_dev))\n",
    "    print('# of test data samples:', len(dataset_test))\n",
    "    print('-'*100)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 165
    },
    "id": "KWh312XMcEFE",
    "outputId": "a8a15912-fd1d-40d5-ed50-be97cb1ae7a0"
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "ignored",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-fa5225a899a7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mN_fingerprints\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'N_fingerprints' is not defined"
     ]
    }
   ],
   "source": [
    "print(N_fingerprints)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "-219K6fOcEHS",
    "outputId": "d270673e-1dea-4696-81c0-7566bf7330a4"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating a model.\n",
      "# of model parameters: 180251\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Start training.\n",
      "The result is saved in the output directory every epoch!\n"
     ]
    }
   ],
   "source": [
    "dim = 50\n",
    "iteration=100\n",
    "print('Creating a model.')\n",
    "torch.manual_seed(1234)\n",
    "model = MolecularGraphNeuralNetwork(\n",
    "        N_fingerprints, dim, layer_hidden, layer_output).to(device)\n",
    "trainer = Trainer(model)\n",
    "tester = Tester(model)\n",
    "print('# of model parameters:',\n",
    "      sum([np.prod(p.size()) for p in model.parameters()]))\n",
    "print('-'*100)\n",
    "\n",
    "#     file_result = '../output/result--' + setting + '.txt'\n",
    "file_result = train_data_path + '/setting1_run.txt'\n",
    "\n",
    "if task == 'classification':\n",
    "    result = 'Epoch\\tTime(sec)\\tLoss_train\\tAUC_dev\\tAUC_test'\n",
    "if task == 'regression':\n",
    "    result = 'Epoch\\tTime(sec)\\tLoss_train\\tMAE_dev\\tMAE_test'\n",
    "\n",
    "with open(file_result, 'w') as f:\n",
    "    f.write(result + '\\n')\n",
    "\n",
    "print('Start training.')\n",
    "  print('The result is saved in the output directory every epoch!')\n",
    "\n",
    "all_loss_train = []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "_vGT6ZBScEJl",
    "outputId": "a2e36a95-5051-42c6-8994-ba3c36b7b99c"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\t29.9315525539987\t551.2848766446114\t0.5832435384504295\t0.622982444260747\n",
      "The training will finish in about 1 hours 39 minutes.\n",
      "----------------------------------------------------------------------------------------------------\n",
      "0\t29.9315525539987\t551.2848766446114\t0.5832435384504295\t0.622982444260747\n",
      "1\t59.74462474199936\t482.29436922073364\t0.5516031286704355\t0.5940023453896627\n",
      "2\t93.15232365499833\t470.91934306174517\t0.5515762088119903\t0.5941910936845801\n",
      "3\t123.22701256599976\t468.1806314140558\t0.5469057483271232\t0.5877234701239146\n",
      "4\t152.79523120799968\t463.9287303239107\t0.5521752930273268\t0.5953469445155217\n",
      "5\t183.2421351239991\t461.8805976808071\t0.5517123117581171\t0.5960749223828316\n",
      "6\t212.86336595600005\t462.88788129389286\t0.5507626289416746\t0.5937085735148344\n",
      "7\t243.66531294299966\t460.2061123922467\t0.5461046916236538\t0.5854636483257397\n",
      "8\t273.0783623309999\t457.9523325264454\t0.5471832461508378\t0.5889735054702331\n",
      "9\t302.35882433199913\t457.5110407471657\t0.546294561098524\t0.589263603402636\n",
      "10\t331.2660439659994\t456.0149440765381\t0.5471245226997536\t0.5903333809264959\n",
      "11\t360.7909588479997\t453.491761431098\t0.5432721708417049\t0.5895529631659007\n",
      "12\t389.8546316869997\t452.6796629726887\t0.5411478137477654\t0.584530004285849\n",
      "13\t419.10335836099875\t452.448425590992\t0.5400831376137939\t0.5808025860327941\n",
      "14\t448.1588411019984\t449.57342906296253\t0.5407835069383601\t0.5807619788621863\n",
      "15\t477.8665656180001\t447.5326173081994\t0.5374204620678265\t0.5837858647509263\n",
      "16\t507.0861924309993\t445.36897680163383\t0.5400450222722721\t0.5780388970711292\n",
      "17\t536.1184736029991\t440.85564194619656\t0.5342028445043456\t0.5775262985579096\n",
      "18\t564.6258411729996\t437.2208852916956\t0.5363225640480351\t0.5794813307317404\n",
      "19\t594.924720044999\t435.90524058043957\t0.5303770923702651\t0.573732434696494\n",
      "20\t626.4562931579985\t431.6561325341463\t0.5277312396374926\t0.5677037641616204\n",
      "21\t659.3782259629988\t429.41383323073387\t0.5294521438377595\t0.5696440173656895\n",
      "22\t689.1885846199984\t428.21405106782913\t0.5379555333720288\t0.5719491508373847\n",
      "23\t719.137920363999\t425.7470615655184\t0.534920077342092\t0.5764175153609652\n",
      "24\t748.7122715589994\t423.2017601579428\t0.5233598000063183\t0.5618248386079302\n",
      "25\t778.7271913299992\t422.46977593004704\t0.5231300515946135\t0.5637552880667723\n",
      "26\t808.4535626009983\t420.81533670425415\t0.5361309634720716\t0.5748329337609884\n",
      "27\t838.2189166569988\t420.3884961903095\t0.5349640684672318\t0.5732329150365714\n",
      "28\t867.5014978179988\t418.6010890305042\t0.5238357897089848\t0.5650706237368286\n",
      "29\t897.2291226909983\t417.41991294920444\t0.5261595559205401\t0.5643818014095991\n",
      "30\t926.5196966029998\t416.9034711420536\t0.5291900688534987\t0.5674838944887504\n",
      "31\t956.2091206469995\t416.30038022994995\t0.5188745107551123\t0.5624282782574972\n",
      "32\t985.4671381119988\t415.8383354097605\t0.5185570079306572\t0.56100927978563\n",
      "33\t1014.5701686550001\t414.47491732239723\t0.5238940422434614\t0.5638980740690843\n",
      "34\t1043.3930237389995\t413.77462743222713\t0.5196012762531996\t0.563950671247947\n",
      "35\t1072.651018339\t413.1325785443187\t0.5255093828289896\t0.5685983929601617\n",
      "36\t1101.713066586999\t413.18450866639614\t0.5207579091107166\t0.5649984899564431\n",
      "37\t1131.1481995319991\t412.66722835600376\t0.5212149890226567\t0.5678088814239853\n",
      "38\t1159.9664345399997\t411.83250617980957\t0.5168049025189755\t0.5590243680966206\n",
      "39\t1189.7392898499984\t412.44381215423346\t0.5169146725031006\t0.5578833257540678\n",
      "40\t1219.1415785069985\t411.165355630219\t0.5188198589688285\t0.5616010808447996\n",
      "41\t1248.8961652929993\t410.85041937977076\t0.5563102271038651\t0.5935974852836284\n",
      "42\t1278.0156382349996\t411.9445369541645\t0.5229874209363231\t0.5611363378797586\n",
      "43\t1308.5627363709991\t409.50546012818813\t0.530365171207222\t0.5703894077871855\n",
      "44\t1339.062849971\t409.6966353058815\t0.5149851045875308\t0.5589885982813744\n",
      "45\t1368.563092335\t409.28913051635027\t0.5192904824872899\t0.5625024574889969\n",
      "46\t1397.3771198719987\t408.63487569242716\t0.5179531037378609\t0.560864961103369\n",
      "47\t1427.0116506179984\t410.2792048379779\t0.5188077267617807\t0.5573321374677694\n",
      "48\t1456.3917313539987\t408.87269726395607\t0.5150836048925284\t0.5608032191086274\n",
      "49\t1485.572481366\t408.07426120340824\t0.5177239612216714\t0.5564876618637489\n",
      "50\t1514.425466314\t407.49732822179794\t0.5168604097705358\t0.5604489768563937\n",
      "51\t1544.3665817009987\t407.0355277955532\t0.5160244734666118\t0.557162417184848\n",
      "52\t1573.5607041209987\t407.85931445658207\t0.5170845962044951\t0.563102631654877\n",
      "53\t1603.155564585999\t407.5275691449642\t0.5178268950740694\t0.5595315341192942\n",
      "54\t1632.432761266\t407.1138945221901\t0.5133034384136889\t0.5568730625978265\n",
      "55\t1662.1313379539988\t407.62705190479755\t0.5121480823367849\t0.5540475182999403\n",
      "56\t1690.981503253999\t406.30230912566185\t0.5166487595124493\t0.556689630613113\n",
      "57\t1720.3420625709987\t406.7234820872545\t0.524771103553591\t0.5676715433645325\n",
      "58\t1749.8901651039996\t405.6118063181639\t0.512955039089466\t0.5537194759971821\n",
      "59\t1780.0785128049993\t406.2021546214819\t0.51305865729662\t0.5554900042598064\n",
      "60\t1809.4644247469987\t405.71920067071915\t0.5172097485939523\t0.5591459781523699\n",
      "61\t1838.968988576\t405.4282493889332\t0.5134690435585887\t0.5520662894577553\n",
      "62\t1867.8715382779992\t406.07865446805954\t0.5100089620947874\t0.5520534059405326\n",
      "63\t1897.2548644119997\t405.7290844768286\t0.5127833570076165\t0.5558122030951274\n",
      "64\t1926.0907051909999\t407.6423799097538\t0.5194012754001034\t0.5638332933894334\n",
      "65\t1955.3438565359993\t405.2487693578005\t0.5162254566158486\t0.5541057651872054\n",
      "66\t1984.117712362\t405.4512250870466\t0.5142543100768853\t0.561827462605941\n",
      "67\t2013.783093738999\t405.8037545979023\t0.5136035339994595\t0.5534241477705729\n",
      "68\t2043.1916996299988\t404.8315854370594\t0.5139551337526733\t0.5544673304947523\n",
      "69\t2072.9982845619998\t404.67292089015245\t0.5134688053905568\t0.5570007571090873\n",
      "70\t2102.9010055159997\t405.0330130904913\t0.5142995008854053\t0.5532102965248319\n",
      "71\t2132.300502406999\t404.95346212387085\t0.5119764404478824\t0.5543457034172921\n",
      "72\t2161.313900470999\t404.4219886511564\t0.5147846481238415\t0.5607982499037798\n",
      "73\t2190.5804734489993\t404.79492346942425\t0.5124250196723316\t0.5557095945798434\n",
      "74\t2219.6639813089987\t404.69202745705843\t0.515609550923036\t0.5589267522392747\n",
      "75\t2250.066504168999\t404.31149384379387\t0.5144800486831314\t0.5573070151511675\n",
      "76\t2279.243821684\t404.45515267550945\t0.510292575355376\t0.5558791523961685\n",
      "77\t2308.8723200489985\t403.36073353886604\t0.5114189668842689\t0.5563625014114838\n",
      "78\t2338.052197245999\t403.78413907438517\t0.5133432596951639\t0.5596844955266286\n",
      "79\t2367.4314312339993\t403.50090508162975\t0.5227596669091633\t0.5706276239817724\n",
      "80\t2396.854314623999\t403.83490954339504\t0.5156989112216568\t0.5564253363758326\n",
      "81\t2427.9339927269993\t403.93937653303146\t0.5149477841776042\t0.5564801097680361\n",
      "82\t2458.6923905699987\t403.84545300900936\t0.5115149970679367\t0.5511494262650227\n",
      "83\t2490.0924317869994\t403.4301765561104\t0.515008564832244\t0.555710228637625\n",
      "84\t2521.2913909599993\t404.02193807065487\t0.518397858971707\t0.5579592212232259\n",
      "85\t2552.630477662\t403.44380632042885\t0.5105817538585969\t0.5565164598402305\n",
      "86\t2582.139947246\t403.23605654388666\t0.5186131345862537\t0.5611349634405894\n",
      "87\t2611.7715023809997\t403.8478438705206\t0.5162737010431805\t0.555030993815416\n",
      "88\t2640.9701518989987\t402.67946011573076\t0.5094592697345997\t0.5534501375047824\n",
      "89\t2670.2204221009997\t404.5767894014716\t0.5154906736834731\t0.5557760338981946\n",
      "90\t2699.3353704579986\t403.2983004748821\t0.5145450419182608\t0.5578781141035067\n",
      "91\t2728.9901001979997\t402.2684445530176\t0.514953468241818\t0.5578881658078768\n",
      "92\t2758.0930555219984\t403.5679841339588\t0.5094567321032417\t0.553713858985366\n",
      "93\t2787.290682269999\t402.72347985208035\t0.5191305994490966\t0.5601666277054793\n",
      "94\t2816.4516471899988\t402.9489421695471\t0.5172163183422769\t0.5547180898774129\n",
      "95\t2845.9675989049992\t403.16738104075193\t0.5108504950415909\t0.5509529347488513\n",
      "96\t2875.013506187999\t402.4647255539894\t0.5169866125175485\t0.5549270618993503\n",
      "97\t2904.7378477149996\t402.32327427715063\t0.5100323358145129\t0.5553314641404611\n",
      "98\t2934.292482113\t403.2235030606389\t0.5119479137647003\t0.5503269542887425\n",
      "99\t2963.642291778\t402.5515436977148\t0.5126965706362163\t0.5543820230032389\n"
     ]
    }
   ],
   "source": [
    "# np.random.seed(1234)\n",
    "\n",
    "start = timeit.default_timer()\n",
    "\n",
    "for epoch in range(iteration):\n",
    "\n",
    "    # Dont decay lr\n",
    "    # if epoch % decay_interval == 0:\n",
    "    #     trainer.optimizer.param_groups[0]['lr'] *= lr_decay\n",
    "\n",
    "    loss_train = trainer.train(dataset_train)\n",
    "\n",
    "    if task == 'regression':\n",
    "        prediction_dev = tester.test_regressor(dataset_dev)\n",
    "        prediction_test = tester.test_regressor(dataset_test)\n",
    "\n",
    "    time = timeit.default_timer() - start\n",
    "\n",
    "    if epoch == 1:\n",
    "        minutes = time * iteration / 60\n",
    "        hours = int(minutes / 60)\n",
    "        minutes = int(minutes - 60 * hours)\n",
    "        print('The training will finish in about',\n",
    "              hours, 'hours', minutes, 'minutes.')\n",
    "        print('-'*100)\n",
    "        print(result)\n",
    "\n",
    "    result = '\\t'.join(map(str, [epoch, time, loss_train,\n",
    "                                 prediction_dev, prediction_test]))\n",
    "    tester.save_result(result, file_result)\n",
    "\n",
    "    print(result)\n",
    "    all_loss_train.append(loss_train)\n",
    "    epoch += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "3D3vcKwZjGyA",
    "outputId": "467c18fe-6273-4431-dd3d-70ba777861ba"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_loss_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Tl4uRRkajG1c"
   },
   "outputs": [],
   "source": [
    "# torch.save(model.state_dict(), train_data_path +'/model_100iter_r3_scaled_colab_hr1')\n",
    "#Later to restore:\n",
    "aaaa = model.load_state_dict(torch.load(train_data_path +'/model_100iter_r3_scaled_colab_hr1.h5'))\n",
    "# model.eval()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "rrzsYdLNEgWC"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0DItGMbecEMG",
    "outputId": "b806f40e-7528-4e8d-b785-cb5b3b43960e"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting torchinfo\n",
      "  Downloading torchinfo-1.6.5-py3-none-any.whl (21 kB)\n",
      "Installing collected packages: torchinfo\n",
      "Successfully installed torchinfo-1.6.5\n"
     ]
    }
   ],
   "source": [
    "# !pip install hiddenlayer\n",
    "!pip install torchinfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "s1B8XgFJcEOz",
    "outputId": "b2fd95fb-e6bf-4973-c567-a6684befc9e8"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 9166\n",
      "The training will finish in about 0 hours 1 minutes.\n",
      "----------------------------------------------------------------------------------------------------\n",
      "0\t33.53304521900009\t242657.7946395874\t12.522179454550418\t12.134580989981758\n",
      "1\t25.38518947099965\t117509.69805145264\t12.588177393041057\t12.431383783777843\n",
      "9166 18332\n",
      "2\t43.331842485999914\t124835.9345703125\t12.743936663626103\t12.406385482467071\n",
      "18332 27498\n",
      "3\t46.3504230210001\t0\t12.743936663626103\t12.406385482467071\n"
     ]
    }
   ],
   "source": [
    "# np.random.seed(1234)\n",
    "\n",
    "# start = timeit.default_timer()\n",
    "\n",
    "# epoch_batch_size = int( len(dataset_train)/iteration )\n",
    "# batch_start_nd = 0\n",
    "# batch_end_nd = epoch_batch_size\n",
    "\n",
    "# iteration = 3\n",
    "\n",
    "# for epoch in range(iteration):\n",
    "#     print(batch_start_nd, batch_end_nd)\n",
    "#     epoch += 1\n",
    "#     if epoch % decay_interval == 0:\n",
    "#         trainer.optimizer.param_groups[0]['lr'] *= lr_decay\n",
    "\n",
    "#     if(epoch != iteration - 1):\n",
    "#       loss_train = trainer.train(dataset_train[batch_start_nd : batch_end_nd])\n",
    "\n",
    "#     if(epoch == iteration - 1):\n",
    "#       loss_train = trainer.train(dataset_train[batch_start_nd :])\n",
    "\n",
    "#     if task == 'regression':\n",
    "#         prediction_dev = tester.test_regressor(dataset_dev)\n",
    "#         prediction_test = tester.test_regressor(dataset_test)\n",
    "\n",
    "#     time = timeit.default_timer() - start\n",
    "\n",
    "#     if epoch == 1:\n",
    "#         minutes = time * iteration / 60\n",
    "#         hours = int(minutes / 60)\n",
    "#         minutes = int(minutes - 60 * hours)\n",
    "#         print('The training will finish in about',\n",
    "#               hours, 'hours', minutes, 'minutes.')\n",
    "#         print('-'*100)\n",
    "#         print(result)\n",
    "\n",
    "#     result = '\\t'.join(map(str, [epoch, time, loss_train,\n",
    "#                                  prediction_dev, prediction_test]))\n",
    "#     tester.save_result(result, file_result)\n",
    "\n",
    "#     print(result)\n",
    "#     all_loss_train.append(loss_train)\n",
    "\n",
    "#     batch_start_nd = batch_end_nd\n",
    "#     batch_end_nd = batch_start_nd + epoch_batch_size\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "iyuxLdFDcEQ-"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "XFX0OqbHcEVP"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "BYi6dMxIS5EL"
   },
   "source": [
    "#### updated version:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "RgjXrpJZS763"
   },
   "outputs": [],
   "source": [
    "train_data_path = \"/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN\"\n",
    "main_data = pd.read_csv(train_data_path+'/final_data.csv')\n",
    "print(main_data.shape)\n",
    "main_data.head()\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "data_for_gnn = main_data[['drug_a_structure', 'drug_b_structure', 'cell_line', 'synergy']].copy()\n",
    "data_for_gnn = data_for_gnn.sample(frac=1).reset_index().drop(['index'], 1)\n",
    "\n",
    "scaler = StandardScaler().fit(data_for_gnn['synergy'].values.reshape(-1,1))\n",
    "scaled_ops = scaler.transform(data_for_gnn['synergy'].values.reshape(-1,1))\n",
    "data_for_gnn['synergy'] = scaled_ops\n",
    "data_for_gnn.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "6X33A5vN575c"
   },
   "outputs": [],
   "source": [
    "# ### Importing dataset using their script\n",
    "\n",
    "# import requests\n",
    "# import os\n",
    "# import zlib\n",
    "\n",
    "\n",
    "# def _download_file(response, filename):\n",
    "#     with open(filename, 'wb') as f:\n",
    "#         for chunk in response.iter_content(chunk_size=1024):\n",
    "#             f.write(chunk)\n",
    "\n",
    "\n",
    "# def _download_and_decompress_file(response, filename):\n",
    "#     decompressor = zlib.decompressobj(16 + zlib.MAX_WBITS)\n",
    "#     filename = filename[:-3]\n",
    "#     with open(filename, 'w+') as f:\n",
    "#         while True:\n",
    "#             chunk = response.raw.read(1024)\n",
    "#             if not chunk:\n",
    "#                 break\n",
    "#             string = decompressor.decompress(chunk)\n",
    "#             f.write(string)\n",
    "\n",
    "\n",
    "# def download_datasets(selected_datasets, selected_downloads, decompress=False):\n",
    "#     for dataset, path in selected_datasets:\n",
    "#         if not os.path.exists(dataset):\n",
    "#             os.mkdir(dataset)\n",
    "\n",
    "#         for downloadable in selected_downloads:\n",
    "#             url = 'https://maayanlab.cloud/static/hdfs/harmonizome/data/%s/%s' %\\\n",
    "#                   (path, downloadable)\n",
    "#             response = requests.get(url, stream=True)\n",
    "#             filename = '%s/%s' % (dataset, downloadable)\n",
    "\n",
    "#             # Not every dataset has all downloadables.\n",
    "#             if response.status_code != 200:\n",
    "#                 continue\n",
    "\n",
    "#             if decompress and 'txt.gz' in filename:\n",
    "#                 _download_and_decompress_file(response, filename)\n",
    "#             else:\n",
    "#                 _download_file(response, filename)\n",
    "\n",
    "#         print('%s downloaded.' % dataset)\n",
    "\n",
    "\n",
    "# if __name__ == '__main__':\n",
    "#     # Uncomment a dataset to download it.\n",
    "#     download_datasets([\n",
    "# #         ('Achilles Cell Line Gene Essentiality Profiles', 'achilles'),\n",
    "# #         ('Allen Brain Atlas Adult Human Brain Tissue Gene Expression Profiles', 'brainatlasadulthuman'),\n",
    "# #         ('Allen Brain Atlas Adult Mouse Brain Tissue Gene Expression Profiles', 'brainatlasadultmouse'),\n",
    "# #         ('Allen Brain Atlas Developing Human Brain Tissue Gene Expression Profiles by Microarray', 'brainatlasdevelopmentalhumanmicroarray'),\n",
    "# #         ('Allen Brain Atlas Developing Human Brain Tissue Gene Expression Profiles by RNA-seq', 'brainatlasdevelopmentalhumanrnaseq'),\n",
    "# #         ('Allen Brain Atlas Prenatal Human Brain Tissue Gene Expression Profiles', 'brainatlasprenatalhuman'),\n",
    "# #         ('BIND Biomolecular Interactions', 'bind'),\n",
    "# #         ('BioGPS Cell Line Gene Expression Profiles', 'biogpsnci60'),\n",
    "# #         ('BioGPS Human Cell Type and Tissue Gene Expression Profiles', 'biogpshuman'),\n",
    "# #         ('BioGPS Mouse Cell Type and Tissue Gene Expression Profiles', 'biogpsmouse'),\n",
    "# #         ('BioGRID Protein-Protein Interactions', 'biogrid'),\n",
    "# #         ('Biocarta Pathways', 'biocarta'),\n",
    "#         ('CCLE Cell Line Gene CNV Profiles', 'cclecnv'),\n",
    "#         ('CCLE Cell Line Gene Expression Profiles', 'cclemrna'),\n",
    "#         ('CCLE Cell Line Gene Mutation Profiles', 'cclemut'),\n",
    "# #         ('CHEA Transcription Factor Binding Site Profiles', 'chea'),\n",
    "# #         ('CHEA Transcription Factor Targets', 'cheappi'),\n",
    "# #         ('CMAP Signatures of Differentially Expressed Genes for Small Molecules', 'cmap'),\n",
    "# #         ('COMPARTMENTS Curated Protein Localization Evidence Scores', 'jensencompartmentcurated'),\n",
    "# #         ('COMPARTMENTS Experimental Protein Localization Evidence Scores', 'jensencompartmentexpts'),\n",
    "# #         ('COMPARTMENTS Text-mining Protein Localization Evidence Scores', 'jensencompartmenttextmining'),\n",
    "# #         ('CORUM Protein Complexes', 'corum'),\n",
    "# #         ('COSMIC Cell Line Gene CNV Profiles', 'cosmiccnv'),\n",
    "# #         ('COSMIC Cell Line Gene Mutation Profiles', 'cosmicmut'),\n",
    "# #         ('CTD Gene-Chemical Interactions', 'ctdchemical'),\n",
    "# #         ('CTD Gene-Disease Associations', 'ctddisease'),\n",
    "# #         ('ClinVar SNP-Phenotype Associations', 'clinvar'),\n",
    "# #         ('Combined Pathways Pathways', 'combinedpathways'),\n",
    "# #         ('dbGAP Gene-Trait Associations', 'dbgap'),\n",
    "# #         ('DEPOD Substrates of Phosphatases', 'depod'),\n",
    "# #         ('DIP Protein-Protein Interactions', 'dip'),\n",
    "# #         ('DISEASES Curated Gene-Disease Assocation Evidence Scores', 'jensendiseasecurated'),\n",
    "# #         ('DISEASES Experimental Gene-Disease Assocation Evidence Scores', 'jensendiseaseexpts'),\n",
    "# #         ('DISEASES Text-mining Gene-Disease Assocation Evidence Scores', 'jensendiseasetextmining'),\n",
    "# #         ('DrugBank Drug Targets', 'drugbank'),\n",
    "# #         ('ENCODE Histone Modification Site Profiles', 'encodehm'),\n",
    "# #         ('ENCODE Transcription Factor Binding Site Profiles', 'encodetf'),\n",
    "# #         ('ENCODE Transcription Factor Targets', 'encodetfppi'),\n",
    "# #         ('ESCAPE Omics Signatures of Genes and Proteins for Stem Cells', 'escape'),\n",
    "# #         ('GAD Gene-Disease Associations', 'gad'),\n",
    "# #         ('GAD High Level Gene-Disease Associations', 'gadhighlevel'),\n",
    "# #         ('GDSC Cell Line Gene Expression Profiles', 'gdsc'),\n",
    "# #         ('GEO Signatures of Differentially Expressed Genes for Diseases', 'geodisease'),\n",
    "# #         ('GEO Signatures of Differentially Expressed Genes for Gene Perturbations', 'geogene'),\n",
    "# #         ('GEO Signatures of Differentially Expressed Genes for Kinase Perturbations', 'geokinase'),\n",
    "# #         ('GEO Signatures of Differentially Expressed Genes for Small Molecules', 'geochemical'),\n",
    "# #         ('GEO Signatures of Differentially Expressed Genes for Transcription Factor Perturbations', 'geotf'),\n",
    "# #         ('GEO Signatures of Differentially Expressed Genes for Viral Infections', 'geovirus'),\n",
    "# #         ('GO Biological Process Annotations', 'gobp'),\n",
    "# #         ('GO Cellular Component Annotations', 'gocc'),\n",
    "# #         ('GO Molecular Function Annotations', 'gomf'),\n",
    "# #         ('GTEx Tissue Gene Expression Profiles', 'gtextissue'),\n",
    "# #         ('GTEx Tissue Sample Gene Expression Profiles', 'gtexsample'),\n",
    "# #         ('GTEx eQTL', 'gtexeqtl'),\n",
    "# #         ('GWAS Catalog SNP-Phenotype Associations', 'gwascatalog'),\n",
    "# #         ('GWASdb SNP-Disease Associations', 'gwasdbdisease'),\n",
    "# #         ('GWASdb SNP-Phenotype Associations', 'gwasdbphenotype'),\n",
    "# #         ('GeneRIF Biological Term Annotations', 'generif'),\n",
    "# #         ('GeneSigDB Published Gene Signatures', 'genesigdb'),\n",
    "# #         ('Graph of Medicine EHR Text-mining Clinical Term Annotations', 'graphofmedicine'),\n",
    "# #         ('Guide to Pharmacology Chemical Ligands of Receptors', 'guidetopharmchemical'),\n",
    "# #         ('Guide to Pharmacology Protein Ligands of Receptors', 'guidetopharmprotein'),\n",
    "# #         ('HMDB Metabolites of Enzymes', 'hmdb'),\n",
    "# #         ('HPA Cell Line Gene Expression Profiles', 'hpacelllines'),\n",
    "# #         ('HPA Tissue Gene Expression Profiles', 'hpatissuesmrna'),\n",
    "# #         ('HPA Tissue Protein Expression Profiles', 'hpatissuesprotein'),\n",
    "# #         ('HPA Tissue Sample Gene Expression Profiles', 'hpasamples'),\n",
    "# #         ('HPM Cell Type and Tissue Protein Expression Profiles', 'hpm'),\n",
    "# #         ('HPO Gene-Disease Associations', 'hpo'),\n",
    "# #         ('HPRD Protein-Protein Interactions', 'hprd'),\n",
    "# #         ('Heiser et al., PNAS, 2011 Cell Line Gene Expression Profiles', 'heiser'),\n",
    "# #         ('HuGE Navigator Gene-Phenotype Associations', 'hugenavigator'),\n",
    "# #         ('Hub Proteins Protein-Protein Interactions', 'hubs'),\n",
    "# #         ('HumanCyc Biomolecular Interactions', 'humancycppi'),\n",
    "# #         ('HumanCyc Pathways', 'humancyc'),\n",
    "# #         ('IntAct Biomolecular Interactions', 'intact'),\n",
    "# #         ('InterPro Predicted Protein Domain Annotations', 'interpro'),\n",
    "# #         ('JASPAR Predicted Transcription Factor Targets', 'jasparpwm'),\n",
    "# #         ('KEA Substrates of Kinases', 'kea'),\n",
    "# #         ('KEGG Biomolecular Interactions', 'keggppi'),\n",
    "# #         ('KEGG Pathways', 'kegg'),\n",
    "# #         ('Kinativ Kinase Inhibitor Bioactivity Profiles', 'kinativ'),\n",
    "# #         ('KinomeScan Kinase Inhibitor Targets', 'kinomescan'),\n",
    "# #         ('Klijn et al., Nat. Biotechnol., 2015 Cell Line Gene CNV Profiles', 'klijncnv'),\n",
    "# #         ('Klijn et al., Nat. Biotechnol., 2015 Cell Line Gene Expression Profiles', 'klijnmrna'),\n",
    "# #         ('Klijn et al., Nat. Biotechnol., 2015 Cell Line Gene Mutation Profiles', 'klijnmut'),\n",
    "# #         ('LINCS L1000 CMAP Signatures of Differentially Expressed Genes for Gene Knockdowns', 'lincscmapgene'),\n",
    "# #         ('LINCS L1000 CMAP Signatures of Differentially Expressed Genes for Small Molecules', 'lincscmapchemical'),\n",
    "# #         ('LOCATE Curated Protein Localization Annotations', 'locate'),\n",
    "# #         ('LOCATE Predicted Protein Localization Annotations', 'locatepredicted'),\n",
    "# #         ('MPO Gene-Phenotype Associations', 'mgimpo'),\n",
    "# #         ('MSigDB Cancer Gene Co-expression Modules', 'msigdbcomp'),\n",
    "# #         ('MSigDB Signatures of Differentially Expressed Genes for Cancer Gene Perturbations', 'msigdbonc'),\n",
    "# #         ('MiRTarBase microRNA Targets', 'mirtarbase'),\n",
    "# #         ('MotifMap Predicted Transcription Factor Targets', 'motifmap'),\n",
    "# #         ('NURSA Protein Complexes', 'nursa'),\n",
    "# #         ('NURSA Protein-Protein Interactions', 'nursappi'),\n",
    "# #         ('OMIM Gene-Disease Associations', 'omim'),\n",
    "# #         ('PANTHER Biomolecular Interactions', 'pantherppi'),\n",
    "# #         ('PANTHER Pathways', 'panther'),\n",
    "# #         ('PID Biomolecular Interactions', 'pidppi'),\n",
    "# #         ('PID Pathways', 'pid'),\n",
    "# #         ('Pathway Commons Protein-Protein Interactions', 'pc'),\n",
    "# #         ('PhosphoSitePlus Phosphosite-Disease Associations', 'phosphositeplusdisease'),\n",
    "# #         ('PhosphoSitePlus Substrates of Kinases', 'phosphositeplus'),\n",
    "# #         ('Phosphosite Textmining Biological Term Annotations', 'phosphositetextmining'),\n",
    "# #         ('ProteomicsDB Cell Type and Tissue Protein Expression Profiles', 'proteomicsdb'),\n",
    "# #         ('Reactome Biomolecular Interactions', 'reactomeppi'),\n",
    "# #         ('Reactome Pathways', 'reactome'),\n",
    "# #         ('Recon X Predicted Biomolecular Interactions', 'reconx'),\n",
    "# #         ('Roadmap Epigenomics Cell and Tissue DNA Accessibility Profiles', 'epigenomicsdnaaccessibility'),\n",
    "# #         ('Roadmap Epigenomics Cell and Tissue DNA Methylation Profiles', 'epigenomicsdnamethylation'),\n",
    "# #         ('Roadmap Epigenomics Cell and Tissue Gene Expression Profiles', 'epigenomicsmrna'),\n",
    "# #         ('Roadmap Epigenomics Histone Modification Site Profiles', 'epigenomicshm'),\n",
    "# #         ('SILAC Phosphoproteomics Signatures of Differentially Phosphorylated Proteins for Drugs', 'silacdrug'),\n",
    "# #         ('SILAC Phosphoproteomics Signatures of Differentially Phosphorylated Proteins for Gene Perturbations', 'silacgene'),\n",
    "# #         ('SILAC Phosphoproteomics Signatures of Differentially Phosphorylated Proteins for Protein Ligands', 'silacligand'),\n",
    "# #         ('SNPedia SNP-Phenotype Associations', 'snpedia'),\n",
    "# #         ('TCGA Signatures of Differentially Expressed Genes for Tumors', 'tcga'),\n",
    "# #         ('TISSUES Curated Tissue Protein Expression Evidence Scores', 'jensentissuecurated'),\n",
    "# #         ('TISSUES Experimental Tissue Protein Expression Evidence Scores', 'jensentissueexpts'),\n",
    "# #         ('TISSUES Text-mining Tissue Protein Expression Evidence Scores', 'jensentissuetextmining'),\n",
    "# #         ('TRANSFAC Curated Transcription Factor Targets', 'transfac'),\n",
    "# #         ('TRANSFAC Predicted Transcription Factor Targets', 'transfacpwm'),\n",
    "# #         ('TargetScan Predicted Conserved microRNA Targets', 'targetscan'),\n",
    "# #         ('TargetScan Predicted Nonconserved microRNA Targets', 'targetscannonconserved'),\n",
    "# #         ('Virus MINT Protein-Viral Protein Interactions', 'virusmintppi'),\n",
    "# #         ('Virus MINT Protein-Virus Interactions', 'virusmint'),\n",
    "# #         ('Wikipathways Pathways', 'wikipathways'),\n",
    "#     ], [\n",
    "#         'gene_attribute_matrix.txt.gz',\n",
    "#         'gene_attribute_edges.txt.gz',\n",
    "# #         'gene_set_library_crisp.txt.gz',\n",
    "# #         'gene_set_library_up_crisp.txt.gz',\n",
    "# #         'gene_set_library_dn_crisp.txt.gz',\n",
    "# #         'attribute_set_library_crisp.txt.gz',\n",
    "# #         'attribute_set_library_up_crisp.txt.gz',\n",
    "# #         'attribute_set_library_dn_crisp.txt.gz',\n",
    "# #         'gene_similarity_matrix_cosine.txt.gz',\n",
    "# #         'attribute_similarity_matrix_cosine.txt.gz',\n",
    "# #         'gene_list_terms.txt.gz',\n",
    "# #         'attribute_list_entries.txt.gz',\n",
    "#         'processing_script.m'\n",
    "#     ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "VEzaiYgV577h"
   },
   "outputs": [],
   "source": [
    "# main_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ozSjdBhK58l9"
   },
   "source": [
    "#### Pre-processing gene expression attribute matrix and edge matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "h2vcCgR_wKQA",
    "outputId": "a7881420-9ca9-4164-8981-7605db8bd3a4"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18027, 1040)\n"
     ]
    }
   ],
   "source": [
    "ccle_attr = pd.read_csv('/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN/CCLE Cell Line Gene Expression Profiles/gene_attribute_matrix.txt.gz', sep='\\t')\n",
    "print(ccle_attr.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 334
    },
    "id": "MDu6ZD09wKag",
    "outputId": "46e9a45f-8797-41a3-bd02-b2812b21564a"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-db2d468b-a313-46bd-8966-31c11c0c85e5\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#</th>\n",
       "      <th>#.1</th>\n",
       "      <th>CellLine</th>\n",
       "      <th>CHL1</th>\n",
       "      <th>HMCB</th>\n",
       "      <th>HS852T</th>\n",
       "      <th>HS695T</th>\n",
       "      <th>A101D</th>\n",
       "      <th>HS294T</th>\n",
       "      <th>SNU466</th>\n",
       "      <th>...</th>\n",
       "      <th>HEL9217</th>\n",
       "      <th>HEL</th>\n",
       "      <th>UT7</th>\n",
       "      <th>SET2</th>\n",
       "      <th>MOLM16</th>\n",
       "      <th>KU812</th>\n",
       "      <th>TF1</th>\n",
       "      <th>MEG01</th>\n",
       "      <th>KYO1</th>\n",
       "      <th>K562</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>central nervous system</td>\n",
       "      <td>...</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneSym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GeneID/NA</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>na</td>\n",
       "      <td>2200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ITGBL1</td>\n",
       "      <td>na</td>\n",
       "      <td>9358</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LRP1</td>\n",
       "      <td>na</td>\n",
       "      <td>4035</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  1040 columns</p>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-db2d468b-a313-46bd-8966-31c11c0c85e5')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-db2d468b-a313-46bd-8966-31c11c0c85e5 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-db2d468b-a313-46bd-8966-31c11c0c85e5');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "         #  #.1   CellLine       CHL1       HMCB    HS852T     HS695T  \\\n",
       "0        #    #     Tissue       skin       skin      skin       skin   \n",
       "1  GeneSym  NaN  GeneID/NA         na         na        na         na   \n",
       "2     FBN1   na       2200   0.000000   0.000000  0.000000  -0.000000   \n",
       "3   ITGBL1   na       9358  -0.000000  -0.000000  0.000000  -0.000000   \n",
       "4     LRP1   na       4035  -0.000000  -0.000000  0.000000   0.000000   \n",
       "\n",
       "       A101D     HS294T                  SNU466  ...  \\\n",
       "0       skin       skin  central nervous system  ...   \n",
       "1         na         na                      na  ...   \n",
       "2   0.000000   0.000000                0.000000  ...   \n",
       "3  -0.000000  -0.000000                0.000000  ...   \n",
       "4   0.000000   0.000000                1.000000  ...   \n",
       "\n",
       "                              HEL9217                                 HEL  \\\n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue   \n",
       "1                                  na                                  na   \n",
       "2                           -0.000000                            0.000000   \n",
       "3                            0.000000                            0.000000   \n",
       "4                           -0.000000                           -0.000000   \n",
       "\n",
       "                                  UT7                                SET2  \\\n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue   \n",
       "1                                  na                                  na   \n",
       "2                            0.000000                            0.000000   \n",
       "3                            0.000000                            0.000000   \n",
       "4                           -0.000000                           -0.000000   \n",
       "\n",
       "                               MOLM16                               KU812  \\\n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue   \n",
       "1                                  na                                  na   \n",
       "2                            0.000000                            0.000000   \n",
       "3                            0.000000                            0.000000   \n",
       "4                           -0.000000                           -0.000000   \n",
       "\n",
       "                                  TF1                               MEG01  \\\n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue   \n",
       "1                                  na                                  na   \n",
       "2                           -0.000000                            0.000000   \n",
       "3                            0.000000                            0.000000   \n",
       "4                           -0.000000                           -0.000000   \n",
       "\n",
       "                                 KYO1                                K562  \n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue  \n",
       "1                                  na                                  na  \n",
       "2                           -0.000000                           -0.000000  \n",
       "3                            0.000000                           -0.000000  \n",
       "4                           -0.000000                           -0.000000  \n",
       "\n",
       "[5 rows x 1040 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_attr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "3kcqY91T6CwX",
    "outputId": "fcae9eed-a6f1-4950-8d79-86a430b14b32"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18027, 1040)\n",
      "   index Cell_lines\n",
      "0      0       CHL1\n",
      "1      1       HMCB\n",
      "2      2     HS852T\n",
      "3      3     HS695T\n",
      "4      4      A101D\n",
      "   index   Genes Gene_id\n",
      "0      0    FBN1    2200\n",
      "1      1  ITGBL1    9358\n",
      "2      2    LRP1    4035\n",
      "3      3   LTBP2    4053\n",
      "4      4   PARVA   55742\n",
      "                                          expression\n",
      "0  [0.000000, -0.000000, -0.000000, -0.000000, -0...\n",
      "1  [0.000000, -0.000000, -0.000000, -0.000000, -0...\n",
      "2  [0.000000, 0.000000, 0.000000, -0.000000, 0.00...\n",
      "3  [-0.000000, -0.000000, 0.000000, 0.000000, 0.0...\n",
      "4  [0.000000, -0.000000, 0.000000, -0.000000, 0.0...\n"
     ]
    }
   ],
   "source": [
    "# Read attribute matrix\n",
    "ccle_attr = pd.read_csv('/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN/CCLE Cell Line Gene Expression Profiles/gene_attribute_matrix.txt.gz', sep='\\t')\n",
    "print(ccle_attr.shape)\n",
    "cell2ind = pd.DataFrame(columns = ['index','Cell_lines'])\n",
    "cell2ind['Cell_lines'] = list(ccle_attr.columns[3:])\n",
    "cell2ind['index'] =  list(cell2ind.index)\n",
    "print(cell2ind.head())\n",
    "gene2ind = pd.DataFrame(columns = ['index','Genes', 'Gene_id'])\n",
    "gene2ind['Genes'] =  list(ccle_attr['#'].values[2:])\n",
    "gene2ind['index'] =  list(gene2ind.index)\n",
    "gene2ind['Gene_id'] =  list(ccle_attr['CellLine'].values[2:])\n",
    "\n",
    "print(gene2ind.head())\n",
    "cell2exp = pd.DataFrame(columns = ['expression'])\n",
    "l = []\n",
    "for cells_ind in range(len(cell2ind['Cell_lines'])):\n",
    "    l.append(list(ccle_attr[cell2ind['Cell_lines'].iloc[cells_ind]].values[2:]))\n",
    "    \n",
    "cell2exp['expression'] = l\n",
    "print(cell2exp.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "id": "q7pGl2A51yzC"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "QXhJbfJUIFML",
    "outputId": "47aadf87-fbf6-45a0-fbc2-e1e42a959765"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18025, 3)\n"
     ]
    }
   ],
   "source": [
    "print(gene2ind.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "id": "487fhQ4huwzh"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 206
    },
    "id": "SM2AQ1Mr6C34",
    "outputId": "78af3660-9c36-41d9-f0a8-b5d325f52fc3"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-88d9f7b3-d38e-4221-a0f1-af76ee10fda1\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein1</th>\n",
       "      <th>Protein2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1394</td>\n",
       "      <td>2778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6331</td>\n",
       "      <td>17999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>122704</td>\n",
       "      <td>54460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2597</td>\n",
       "      <td>2911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4790</td>\n",
       "      <td>79155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-88d9f7b3-d38e-4221-a0f1-af76ee10fda1')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-88d9f7b3-d38e-4221-a0f1-af76ee10fda1 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-88d9f7b3-d38e-4221-a0f1-af76ee10fda1');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "   Protein1  Protein2\n",
       "0      1394      2778\n",
       "1      6331     17999\n",
       "2    122704     54460\n",
       "3      2597      2911\n",
       "4      4790     79155"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pp_int = pd.read_csv('/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN/PP-Pathways_ppi.csv.gz', header=None)\n",
    "pp_int.columns = ['Protein1', 'Protein2']\n",
    "pp_int.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "id": "sWegpWg86C6o"
   },
   "outputs": [],
   "source": [
    "# This dataset has proper gene-gene interactions\n",
    "# pp_int[pp_int['Protein1'] == 2200][:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "C2zolyQ1Ny3L",
    "outputId": "64d824a3-ae8a-4fc0-e9ed-6fcce38e2a90"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6\n"
     ]
    }
   ],
   "source": [
    "# If any genes are cell line ids, remove them\n",
    "cell_ids = cell2ind['Cell_lines'].unique()\n",
    "to_drop = []\n",
    "for i in range(len(gene2ind)):\n",
    "  if(gene2ind.iloc[i]['Genes'] in cell_ids):\n",
    "    to_drop.append(i)\n",
    "\n",
    "print(len(to_drop))\n",
    "\n",
    "gene2ind = gene2ind.drop(to_drop, axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "_piAJkvF87Ws",
    "outputId": "0b527739-d3ba-493a-c6ff-785195b9d897"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15000, 3)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene2ind.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "DZWvDK8h040q",
    "outputId": "fe21dff2-387b-49f0-e6a3-9717c62f7873"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total gene ids in pp_int dataset:  21557\n",
      "Total Common intersecting gene ids: 14994  and Total uncommon non-intersecting gene ids:  3025\n"
     ]
    }
   ],
   "source": [
    "# Check and take only intersection of gene ids\n",
    "\n",
    "# Create set of all values from both p1 and p2 columns\n",
    "p1p2 = set()\n",
    "# for i in range(len(pp_int)):\n",
    "#   p1p2.add(pp_int['Protein1'].iloc[i])\n",
    "#   p1p2.add(pp_int['Protein2'].iloc[i])\n",
    "\n",
    "p1p2 = set(np.concatenate((pp_int['Protein1'].unique(), pp_int['Protein2'].unique())))\n",
    "print('Total gene ids in pp_int dataset: ', len(p1p2))\n",
    "\n",
    "# Check if any gene ids from deepsynergy dataset is missing in this p1p2\n",
    "common_gene_ids = []\n",
    "uncommon_gene_ids = []\n",
    "\n",
    "for i in range(len(gene2ind)):\n",
    "  if(gene2ind.iloc[i]['Gene_id'] in p1p2):\n",
    "    common_gene_ids.append(gene2ind.iloc[i]['Gene_id'])\n",
    "  else:\n",
    "    uncommon_gene_ids.append(gene2ind.iloc[i]['Gene_id'])\n",
    "\n",
    "print('Total Common intersecting gene ids:', len(common_gene_ids), ' and Total uncommon non-intersecting gene ids: ', len(uncommon_gene_ids))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "id": "kT27Oo096DkY"
   },
   "outputs": [],
   "source": [
    "# Take only 5k genes (avoid computational problems )\n",
    "# final_gene_ids = common_gene_ids[:10000]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 334
    },
    "id": "qCsrxrEwd_Td",
    "outputId": "5aef1083-9fda-4ea3-e3c1-2bef23143bd9"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-b129df7f-8a37-4408-bd32-f594bdb6127d\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#</th>\n",
       "      <th>#.1</th>\n",
       "      <th>CellLine</th>\n",
       "      <th>CHL1</th>\n",
       "      <th>HMCB</th>\n",
       "      <th>HS852T</th>\n",
       "      <th>HS695T</th>\n",
       "      <th>A101D</th>\n",
       "      <th>HS294T</th>\n",
       "      <th>SNU466</th>\n",
       "      <th>...</th>\n",
       "      <th>HEL9217</th>\n",
       "      <th>HEL</th>\n",
       "      <th>UT7</th>\n",
       "      <th>SET2</th>\n",
       "      <th>MOLM16</th>\n",
       "      <th>KU812</th>\n",
       "      <th>TF1</th>\n",
       "      <th>MEG01</th>\n",
       "      <th>KYO1</th>\n",
       "      <th>K562</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>skin</td>\n",
       "      <td>central nervous system</td>\n",
       "      <td>...</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "      <td>haematopoietic and lymphoid tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneSym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GeneID/NA</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>na</td>\n",
       "      <td>2200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ITGBL1</td>\n",
       "      <td>na</td>\n",
       "      <td>9358</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LRP1</td>\n",
       "      <td>na</td>\n",
       "      <td>4035</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  1040 columns</p>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b129df7f-8a37-4408-bd32-f594bdb6127d')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-b129df7f-8a37-4408-bd32-f594bdb6127d button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-b129df7f-8a37-4408-bd32-f594bdb6127d');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "         #  #.1   CellLine       CHL1       HMCB    HS852T     HS695T  \\\n",
       "0        #    #     Tissue       skin       skin      skin       skin   \n",
       "1  GeneSym  NaN  GeneID/NA         na         na        na         na   \n",
       "2     FBN1   na       2200   0.000000   0.000000  0.000000  -0.000000   \n",
       "3   ITGBL1   na       9358  -0.000000  -0.000000  0.000000  -0.000000   \n",
       "4     LRP1   na       4035  -0.000000  -0.000000  0.000000   0.000000   \n",
       "\n",
       "       A101D     HS294T                  SNU466  ...  \\\n",
       "0       skin       skin  central nervous system  ...   \n",
       "1         na         na                      na  ...   \n",
       "2   0.000000   0.000000                0.000000  ...   \n",
       "3  -0.000000  -0.000000                0.000000  ...   \n",
       "4   0.000000   0.000000                1.000000  ...   \n",
       "\n",
       "                              HEL9217                                 HEL  \\\n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue   \n",
       "1                                  na                                  na   \n",
       "2                           -0.000000                            0.000000   \n",
       "3                            0.000000                            0.000000   \n",
       "4                           -0.000000                           -0.000000   \n",
       "\n",
       "                                  UT7                                SET2  \\\n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue   \n",
       "1                                  na                                  na   \n",
       "2                            0.000000                            0.000000   \n",
       "3                            0.000000                            0.000000   \n",
       "4                           -0.000000                           -0.000000   \n",
       "\n",
       "                               MOLM16                               KU812  \\\n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue   \n",
       "1                                  na                                  na   \n",
       "2                            0.000000                            0.000000   \n",
       "3                            0.000000                            0.000000   \n",
       "4                           -0.000000                           -0.000000   \n",
       "\n",
       "                                  TF1                               MEG01  \\\n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue   \n",
       "1                                  na                                  na   \n",
       "2                           -0.000000                            0.000000   \n",
       "3                            0.000000                            0.000000   \n",
       "4                           -0.000000                           -0.000000   \n",
       "\n",
       "                                 KYO1                                K562  \n",
       "0  haematopoietic and lymphoid tissue  haematopoietic and lymphoid tissue  \n",
       "1                                  na                                  na  \n",
       "2                           -0.000000                           -0.000000  \n",
       "3                            0.000000                           -0.000000  \n",
       "4                           -0.000000                           -0.000000  \n",
       "\n",
       "[5 rows x 1040 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_attr.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "oeGs8LoyghSx"
   },
   "source": [
    "#### Re-read cell_attr, this time ignore uncommon gene ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "iwGO6wXKhUqq",
    "outputId": "ed5b8fa8-ebea-44e5-dec7-8a31c001c5d5"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dropping  3025  genes\n",
      "(15002, 1040)\n"
     ]
    }
   ],
   "source": [
    "# Remove rows with uncommon gene ids\n",
    "to_drop = []\n",
    "for i in range(2, len(ccle_attr)):\n",
    "  if(ccle_attr.iloc[i]['CellLine'] in uncommon_gene_ids):\n",
    "    to_drop.append(i)\n",
    "\n",
    "print('dropping ', len(to_drop), ' genes')\n",
    "ccle_attr = ccle_attr.drop(to_drop, axis = 0)\n",
    "print(ccle_attr.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "Lj40K3GsiPcX",
    "outputId": "4d80d73f-bbda-42ed-e75a-23fd884b50a2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(342353, 2)\n"
     ]
    }
   ],
   "source": [
    "# Drop from pp_int dataset\n",
    "to_drop = set()\n",
    "for i in range(len(pp_int)):\n",
    "  if(pp_int.iloc[i]['Protein1'] in uncommon_gene_ids):\n",
    "    to_drop.add(i)\n",
    "  if(pp_int.iloc[i]['Protein2'] in uncommon_gene_ids):\n",
    "    to_drop.add(i)\n",
    "\n",
    "to_drop = list(to_drop)\n",
    "pp_int = pp_int.drop(to_drop, axis = 0)\n",
    "print(pp_int.shape)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "id": "1dxx5ifu8mw5"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "id": "fKij4bg3jWen"
   },
   "outputs": [],
   "source": [
    "# Save these two datasets, no need to run again\n",
    "# ccle_attr.to_csv('/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN/ccle_attr_updated.csv', index=False)\n",
    "# pp_int.to_csv('/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN/pp_int_updated.csv', index=False)\n",
    "\n",
    "# ccle_attr = pd.read_csv('/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN/ccle_attr_updated.csv')\n",
    "# pp_int = pd.read_csv('/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN/pp_int_updated.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "DD-OYETUd_V7",
    "outputId": "05eab93a-96a6-48e6-9873-4a5c7b9cd7e9"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   index Cell_lines\n",
      "0      0       CHL1\n",
      "1      1       HMCB\n",
      "2      2     HS852T\n",
      "3      3     HS695T\n",
      "4      4      A101D\n",
      "   index    Genes  Gene_id\n",
      "0      0      LBH    81606\n",
      "1      1     GLI2     2736\n",
      "2      2    PAPPA     5069\n",
      "3      3  SLC30A4     7782\n",
      "4      4    JAZF1   221895\n",
      "                                          expression\n",
      "0  [-0.0, 0.0, 0.0, 0.0, 0.0, -0.0, -0.0, -0.0, -...\n",
      "1  [-0.0, 1.0, -0.0, 0.0, 0.0, -0.0, 0.0, 0.0, -0...\n",
      "2  [0.0, 0.0, 0.0, 0.0, 0.0, -0.0, -0.0, 0.0, 0.0...\n",
      "3  [-0.0, 0.0, -0.0, 0.0, 0.0, 0.0, -0.0, -0.0, -...\n",
      "4  [-0.0, 0.0, 0.0, -0.0, 0.0, 0.0, 0.0, 0.0, 0.0...\n"
     ]
    }
   ],
   "source": [
    "\n",
    "cell2ind = pd.DataFrame(columns = ['index','Cell_lines'])\n",
    "cell2ind['Cell_lines'] = list(ccle_attr.columns[3:])\n",
    "cell2ind['index'] =  list(cell2ind.index)\n",
    "print(cell2ind.head())\n",
    "gene2ind = pd.DataFrame(columns = ['index','Genes', 'Gene_id'])\n",
    "gene2ind['Genes'] =  list(ccle_attr['#'].values[2:])\n",
    "gene2ind['index'] =  list(gene2ind.index)\n",
    "gene2ind['Gene_id'] =  list(ccle_attr['CellLine'].values[2:])\n",
    "\n",
    "print(gene2ind.head())\n",
    "cell2exp = pd.DataFrame(columns = ['expression'])\n",
    "l = []\n",
    "for cells_ind in range(len(cell2ind['Cell_lines'])):\n",
    "    l.append(list(ccle_attr[cell2ind['Cell_lines'].iloc[cells_ind]].values[2:]))\n",
    "    \n",
    "cell2exp['expression'] = l\n",
    "print(cell2exp.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "id": "8JZvQlfIgoik"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HazTV5MVgokl",
    "outputId": "9cbeda09-3474-4ba9-b881-d2796514b35d"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15000, 3)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene2ind.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 206
    },
    "id": "Q3sUgWGZgo0P",
    "outputId": "018b9f59-6b88-422c-93cc-593a7c068636"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-2f541450-5508-4e35-b044-6462625c854d\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug_a_structure</th>\n",
       "      <th>drug_b_structure</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>synergy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COc1cccc2c1C(=O)c1c(O)c3c(c(O)c1C2=O)CC(O)(C(=...</td>\n",
       "      <td>CCN(CC)CCNC(=O)c1c(C)[nH]c(C=C2C(=O)Nc3ccc(F)c...</td>\n",
       "      <td>UWB1289</td>\n",
       "      <td>-0.353791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C=CCn1c(=O)c2cnc(Nc3ccc(N4CCN(C)CC4)cc3)nc2n1-...</td>\n",
       "      <td>Cn1cc(-c2cnn3c(N)c(Br)c(C4CCCNC4)nc23)cn1</td>\n",
       "      <td>UACC62</td>\n",
       "      <td>1.684066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O=P1(N(CCCl)CCCl)NCCCO1</td>\n",
       "      <td>O=C(CCCCCCC(=O)Nc1ccccc1)NO</td>\n",
       "      <td>UACC62</td>\n",
       "      <td>0.027459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CS(=O)(=O)CCNCc1ccc(-c2ccc3ncnc(Nc4ccc(OCc5ccc...</td>\n",
       "      <td>Cn1c(=O)n(-c2ccc(C(C)(C)C#N)cc2)c2c3cc(-c4cnc5...</td>\n",
       "      <td>NCIH2122</td>\n",
       "      <td>2.094920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>N.N.O=C(O)C1(C(=O)O)CCC1.[Pt]</td>\n",
       "      <td>N#Cc1ccc(Cn2cncc2CN2CCN(c3cccc(Cl)c3)C(=O)C2)cc1</td>\n",
       "      <td>VCAP</td>\n",
       "      <td>-0.066320</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2f541450-5508-4e35-b044-6462625c854d')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-2f541450-5508-4e35-b044-6462625c854d button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-2f541450-5508-4e35-b044-6462625c854d');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "                                    drug_a_structure  \\\n",
       "0  COc1cccc2c1C(=O)c1c(O)c3c(c(O)c1C2=O)CC(O)(C(=...   \n",
       "1  C=CCn1c(=O)c2cnc(Nc3ccc(N4CCN(C)CC4)cc3)nc2n1-...   \n",
       "2                            O=P1(N(CCCl)CCCl)NCCCO1   \n",
       "3  CS(=O)(=O)CCNCc1ccc(-c2ccc3ncnc(Nc4ccc(OCc5ccc...   \n",
       "4                      N.N.O=C(O)C1(C(=O)O)CCC1.[Pt]   \n",
       "\n",
       "                                    drug_b_structure cell_line   synergy  \n",
       "0  CCN(CC)CCNC(=O)c1c(C)[nH]c(C=C2C(=O)Nc3ccc(F)c...   UWB1289 -0.353791  \n",
       "1          Cn1cc(-c2cnn3c(N)c(Br)c(C4CCCNC4)nc23)cn1    UACC62  1.684066  \n",
       "2                        O=C(CCCCCCC(=O)Nc1ccccc1)NO    UACC62  0.027459  \n",
       "3  Cn1c(=O)n(-c2ccc(C(C)(C)C#N)cc2)c2c3cc(-c4cnc5...  NCIH2122  2.094920  \n",
       "4   N#Cc1ccc(Cn2cncc2CN2CCN(c3cccc(Cl)c3)C(=O)C2)cc1      VCAP -0.066320  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell2ind['expression'] = cell2exp['expression'].copy()\n",
    "cell2ind = cell2ind.rename(columns = {'Cell_lines':'cell_line'})\n",
    "cell2ind.head()\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "data_for_gnn_gcn = main_data[['drug_a_structure', 'drug_b_structure', 'cell_line', 'synergy']][:10000].copy()\n",
    "scaler = StandardScaler().fit(data_for_gnn_gcn['synergy'].values.reshape(-1,1))\n",
    "scaled_ops = scaler.transform(data_for_gnn_gcn['synergy'].values.reshape(-1,1))\n",
    "data_for_gnn_gcn['synergy'] = scaled_ops\n",
    "data_for_gnn_gcn = data_for_gnn_gcn.sample(frac=1).reset_index().drop(['index'], 1)\n",
    "data_for_gnn_gcn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "id": "M8YSnCW3Yjfv"
   },
   "outputs": [],
   "source": [
    "l = []\n",
    "\n",
    "for i in range(len(data_for_gnn_gcn)):\n",
    "  cell_name = data_for_gnn_gcn.iloc[i]['cell_line']\n",
    "  # print(cell_name)\n",
    "  try:\n",
    "    expr = np.array(cell2ind[cell2ind['cell_line'] == cell_name]['expression'].values[0], dtype='float32')\n",
    "  except:\n",
    "    expr = [float(0) for j in range(15000)]\n",
    "  l.append(expr)\n",
    "  # new_data.append({ new_data.columns[0]:main_data.iloc[i][new_data.columns[0]], new_data.columns[1]:main_data.iloc[i][new_data.columns[1]], new_data.columns[2]:main_data.iloc[i][new_data.columns[2]], new_data.columns[3]:main_data.iloc[i][new_data.columns[3]] }, ignore_index = True)\n",
    "\n",
    "data_for_gnn_gcn['cell_line_expr'] = l\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "Sy7D9q0NYjh_",
    "outputId": "ed5097d5-cf5f-4a5f-bf96-d77fc899c83d"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "322"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_for_gnn_gcn.head()\n",
    "import gc\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "id": "EOszBusMp989"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "id": "wzar4iLkp-AS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "glAvQgQUAZ5b"
   },
   "source": [
    "#### Calculate adjacency matrix Using the gene protein-protein interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jHJ5OJ2MRrUd"
   },
   "source": [
    "### NOT YET WORKING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 206
    },
    "id": "NC56eqLlLdeK",
    "outputId": "bb9c734d-76dc-4a18-a777-bfdf6887bc87"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-9f24c6c9-51ce-4497-b9e5-a24670d5d944\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein1</th>\n",
       "      <th>Protein2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1394</td>\n",
       "      <td>2778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6331</td>\n",
       "      <td>17999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>122704</td>\n",
       "      <td>54460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2597</td>\n",
       "      <td>2911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4790</td>\n",
       "      <td>79155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-9f24c6c9-51ce-4497-b9e5-a24670d5d944')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-9f24c6c9-51ce-4497-b9e5-a24670d5d944 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-9f24c6c9-51ce-4497-b9e5-a24670d5d944');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "   Protein1  Protein2\n",
       "0      1394      2778\n",
       "1      6331     17999\n",
       "2    122704     54460\n",
       "3      2597      2911\n",
       "4      4790     79155"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pp_int.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "DWjZBDlTM7v-",
    "outputId": "c8c40464-bae3-47fa-ce3a-806f16f221c6"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21557"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Brute-force method ( for some reason sparse_matrixl)\n",
    "# for i in range(len(pp_int)):\n",
    "# len( set(np.concatenate((pp_int['Protein1'].unique(), pp_int['Protein2'].unique()))) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "id": "1F4HqgkdYjkS"
   },
   "outputs": [],
   "source": [
    "# ## Create dummy adjacency matrix for now\n",
    "# import scipy.sparse as sp\n",
    "# adj = sp.coo_matrix((len(data_for_gnn_gcn), len(data_for_gnn_gcn)), dtype=np.float32).toarray()\n",
    "# # type(sp.csr_matrix(adj))\n",
    "\n",
    "# # build symmetric adjacency matrix\n",
    "# # adj = adj + np.matmul(adj.T, adj.T > adj) - np.matmul(adj, adj.T > adj)\n",
    "# # adj\n",
    "# #\n",
    "# def get_adjacency_matrix(df):\n",
    "#   df = pd.crosstab(df.Protein1, df.Protein2)\n",
    "#   idx = df.columns.union(df.index)\n",
    "#   df = df.reindex(index = idx, columns=idx, fill_value=0)\n",
    "\n",
    "# csc = sp.csc_matrix((np.ones_like(df['A']), (df['A'],df['B'])))\n",
    "# csc.toarray()\n",
    "\n",
    "# # Doing this in sparse matrix multiplication\n",
    "# adj_sp = sp.csr_matrix(adj)\n",
    "# adj_sp = adj_sp + adj_sp.T.multiply(adj_sp.T > adj_sp) - adj_sp.multiply(adj_sp.T > adj_sp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "FBBHVsJGLdv6",
    "outputId": "4441a38e-a5b7-4eb0-de40-e63bf6dce810"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21557, 21557)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "G_di = nx.from_pandas_edgelist(pp_int, 'Protein1', 'Protein2', create_using=nx.DiGraph())\n",
    "A_di = nx.to_scipy_sparse_matrix(G_di)\n",
    "A_di.todense().shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "kwGrv9aAL4gB",
    "outputId": "8cb46c1a-4d4d-4531-ef88-585166f3692c"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19341"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pp_int['Protein2'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "xO77PP_KJClq",
    "outputId": "3892e354-4fdd-4fd5-a45d-1f4f3fe8b2ee"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "networkx.classes.graph.Graph"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import networkx as nx\n",
    "G=nx.from_pandas_edgelist(pp_int, 'Protein1', 'Protein2')\n",
    "type(G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OoCZVaazJ1Lx",
    "outputId": "2afcca13-b5dc-4aeb-e879-af28740bfd6e"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21557, 21557)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A = nx.to_scipy_sparse_matrix(G)\n",
    "A.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "g3rZDNfeJ1PX"
   },
   "outputs": [],
   "source": [
    "# pd.crosstab(pp_int.Protein1, pp_int.Protein2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "BDuAqNwhYjmh",
    "outputId": "8500c996-701e-499f-d244-edcca6a12eac"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10000, 10000)"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gc\n",
    "gc.collect()\n",
    "adj_sp.todense().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "id": "hlcO2MNSp8G2"
   },
   "outputs": [],
   "source": [
    "# pp_int_path = '/content/drive/My Drive/Projects and research stuffs/Drug Synergy/Dataset for GNN/Hsapi20170205.mif25.gz'\n",
    "# import defusedxml.ElementTree as ET\n",
    "# ET.parse(pp_int_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "id": "og2864CoL_gF"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "la2jiPTeIvnX"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "ouHs2SYPYjom"
   },
   "outputs": [],
   "source": [
    "\n",
    "def normalize(mx):\n",
    "    \"\"\"Row-normalize sparse matrix\"\"\"\n",
    "    rowsum = np.array(mx.sum(1))\n",
    "    r_inv = np.power(rowsum, -1).flatten()\n",
    "    r_inv[np.isinf(r_inv)] = 0.\n",
    "    r_mat_inv = sp.diags(r_inv)\n",
    "    mx = r_mat_inv.dot(mx)\n",
    "    return mx\n",
    "\n",
    "\n",
    "\n",
    "def accuracy(output, labels):\n",
    "    preds = output.max(1)[1].type_as(labels)\n",
    "    correct = preds.eq(labels).double()\n",
    "    correct = correct.sum()\n",
    "    return correct / len(labels)\n",
    "\n",
    "\n",
    "def sparse_mx_to_torch_sparse_tensor(sparse_mx):\n",
    "    \"\"\"Convert a scipy sparse matrix to a torch sparse tensor.\"\"\"\n",
    "    sparse_mx = sparse_mx.tocoo().astype(np.float32)\n",
    "    indices = torch.from_numpy(\n",
    "        np.vstack((sparse_mx.row, sparse_mx.col)).astype(np.int64))\n",
    "    values = torch.from_numpy(sparse_mx.data)\n",
    "    shape = torch.Size(sparse_mx.shape)\n",
    "    return torch.sparse.FloatTensor(indices, values, shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "-1tD25k8cEXq"
   },
   "outputs": [],
   "source": [
    "#### GNN combined with GCN\n",
    "from collections import defaultdict\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "from rdkit import Chem\n",
    "\n",
    "import torch\n",
    "\n",
    "\n",
    "def create_atoms(mol, atom_dict):\n",
    "    \"\"\"Transform the atom types in a molecule (e.g., H, C, and O)\n",
    "    into the indices (e.g., H=0, C=1, and O=2).\n",
    "    Note that each atom index considers the aromaticity.\n",
    "    \"\"\"\n",
    "    atoms = [a.GetSymbol() for a in mol.GetAtoms()]\n",
    "    for a in mol.GetAromaticAtoms():\n",
    "        i = a.GetIdx()\n",
    "        atoms[i] = (atoms[i], 'aromatic')\n",
    "    atoms = [atom_dict[a] for a in atoms]\n",
    "    return np.array(atoms)\n",
    "\n",
    "\n",
    "def create_ijbonddict(mol, bond_dict):\n",
    "    \"\"\"Create a dictionary, in which each key is a node ID\n",
    "    and each value is the tuples of its neighboring node\n",
    "    and chemical bond (e.g., single and double) IDs.\n",
    "    \"\"\"\n",
    "    i_jbond_dict = defaultdict(lambda: [])\n",
    "    for b in mol.GetBonds():\n",
    "        i, j = b.GetBeginAtomIdx(), b.GetEndAtomIdx()\n",
    "        bond = bond_dict[str(b.GetBondType())]\n",
    "        i_jbond_dict[i].append((j, bond))\n",
    "        i_jbond_dict[j].append((i, bond))\n",
    "    return i_jbond_dict\n",
    "\n",
    "\n",
    "def extract_fingerprints(radius, atoms, i_jbond_dict,\n",
    "                         fingerprint_dict, edge_dict):\n",
    "    \"\"\"Extract the fingerprints from a molecular graph\n",
    "    based on Weisfeiler-Lehman algorithm.\n",
    "    \"\"\"\n",
    "\n",
    "    if (len(atoms) == 1) or (radius == 0):\n",
    "        nodes = [fingerprint_dict[a] for a in atoms]\n",
    "\n",
    "    else:\n",
    "        nodes = atoms\n",
    "        i_jedge_dict = i_jbond_dict\n",
    "\n",
    "        for _ in range(radius):\n",
    "\n",
    "            \"\"\"Update each node ID considering its neighboring nodes and edges.\n",
    "            The updated node IDs are the fingerprint IDs.\n",
    "            \"\"\"\n",
    "            nodes_ = []\n",
    "            for i, j_edge in i_jedge_dict.items():\n",
    "                neighbors = [(nodes[j], edge) for j, edge in j_edge]\n",
    "                fingerprint = (nodes[i], tuple(sorted(neighbors)))\n",
    "                nodes_.append(fingerprint_dict[fingerprint])\n",
    "\n",
    "            \"\"\"Also update each edge ID considering\n",
    "            its two nodes on both sides.\n",
    "            \"\"\"\n",
    "            i_jedge_dict_ = defaultdict(lambda: [])\n",
    "            for i, j_edge in i_jedge_dict.items():\n",
    "                for j, edge in j_edge:\n",
    "                    both_side = tuple(sorted((nodes[i], nodes[j])))\n",
    "                    edge = edge_dict[(both_side, edge)]\n",
    "                    i_jedge_dict_[i].append((j, edge))\n",
    "\n",
    "            nodes = nodes_\n",
    "            i_jedge_dict = i_jedge_dict_\n",
    "\n",
    "    return np.array(nodes)\n",
    "\n",
    "\n",
    "def split_dataset(dataset, ratio):\n",
    "    \"\"\"Shuffle and split a dataset.\"\"\"\n",
    "    # np.random.seed(1234)  # fix the seed for shuffle.\n",
    "    # np.random.shuffle(dataset)\n",
    "    n = int(ratio * len(dataset))\n",
    "    return dataset[:n], dataset[n:]\n",
    "\n",
    "\n",
    "def create_datasets(task, data_for_gnn, radius, device):\n",
    "\n",
    "    dir_dataset =  ''\n",
    "    # print(type(data_for_gnn))\n",
    "    print(data_for_gnn.shape)\n",
    "    \"\"\"Initialize x_dict, in which each key is a symbol type\n",
    "    (e.g., atom and chemical bond) and each value is its index.\n",
    "    \"\"\"\n",
    "    atom_dict = defaultdict(lambda: len(atom_dict))\n",
    "    bond_dict = defaultdict(lambda: len(bond_dict))\n",
    "    fingerprint_dict = defaultdict(lambda: len(fingerprint_dict))\n",
    "    edge_dict = defaultdict(lambda: len(edge_dict))\n",
    "   \n",
    "    \n",
    "    non_working_smiles_index = []\n",
    "    def create_dataset(data):\n",
    "\n",
    "#         print(filename)\n",
    "\n",
    "#         \"\"\"Load a dataset.\"\"\"\n",
    "#         with open(dir_dataset + filename, 'r') as f:\n",
    "#             smiles_property = f.readline().strip().split()\n",
    "#             data_original = f.read().strip().split('\\n')\n",
    "\n",
    "#         \"\"\"Exclude the data contains '.' in its smiles.\"\"\"\n",
    "#         data_original = [data for data in data_original\n",
    "#                          if '.' not in data.split()[0]]\n",
    "\n",
    "        dataset = []\n",
    "        final_index = []\n",
    "\n",
    "        for data_index in range(len(data)):\n",
    "#             print(data_index)\n",
    "            try:\n",
    "        #             smiles, property = data.strip().split()\n",
    "                smiles1, smiles2, property = data.iloc[data_index]['drug_a_structure'], data.iloc[data_index]['drug_b_structure'], data.iloc[data_index]['synergy']\n",
    "\n",
    "                \"\"\"Create each data with the above defined functions.\"\"\"\n",
    "                # for drug A:\n",
    "                mol = Chem.AddHs(Chem.MolFromSmiles(smiles1))\n",
    "                atoms = create_atoms(mol, atom_dict)\n",
    "                molecular_size1 = len(atoms)\n",
    "                i_jbond_dict = create_ijbonddict(mol, bond_dict)\n",
    "                fingerprints1 = extract_fingerprints(radius, atoms, i_jbond_dict,\n",
    "                                                    fingerprint_dict, edge_dict)\n",
    "                adjacency1 = Chem.GetAdjacencyMatrix(mol)\n",
    "\n",
    "                # for drug B:\n",
    "                mol = Chem.AddHs(Chem.MolFromSmiles(smiles2))\n",
    "                atoms = create_atoms(mol, atom_dict)\n",
    "                molecular_size2 = len(atoms)\n",
    "                i_jbond_dict = create_ijbonddict(mol, bond_dict)\n",
    "                fingerprints2 = extract_fingerprints(radius, atoms, i_jbond_dict,\n",
    "                                                    fingerprint_dict, edge_dict)\n",
    "                adjacency2 = Chem.GetAdjacencyMatrix(mol)\n",
    "\n",
    "                \"\"\"Transform the above each data of numpy\n",
    "                to pytorch tensor on a device (i.e., CPU or GPU).\n",
    "                \"\"\"\n",
    "                fingerprints1 = torch.LongTensor(fingerprints1).to(device)\n",
    "                adjacency1 = torch.FloatTensor(adjacency1).to(device)\n",
    "                fingerprints2 = torch.LongTensor(fingerprints2).to(device)\n",
    "                adjacency2 = torch.FloatTensor(adjacency2).to(device)\n",
    "\n",
    "                if task == 'regression':\n",
    "                    property = torch.FloatTensor([[float(property)]]).to(device)\n",
    "\n",
    "                dataset.append((fingerprints1, adjacency1, molecular_size1, fingerprints2, adjacency2, molecular_size2, property))\n",
    "                final_index.append(data_index)\n",
    "            \n",
    "            except:\n",
    "                non_working_smiles_index.append(data_index)\n",
    "            \n",
    "        return dataset, final_index\n",
    "\n",
    "    dataset_train, final_index = create_dataset(data_for_gnn)\n",
    "    print('Skipped '+ str(len(non_working_smiles_index)) + ' smiles in training data')\n",
    "    dataset_train, dataset_dev = split_dataset(dataset_train, 0.9)\n",
    "    # dataset_test = create_dataset(data_for_gnn[5000:])\n",
    "    dataset_test = dataset_dev\n",
    "\n",
    "    N_fingerprints = len(fingerprint_dict)\n",
    "\n",
    "    return dataset_train, dataset_dev, dataset_test, N_fingerprints, final_index\n",
    "\n",
    "\n",
    "\n",
    "import sys\n",
    "import timeit\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch.optim as optim\n",
    "\n",
    "from sklearn.metrics import roc_auc_score\n",
    "\n",
    "# import preprocess as pp\n",
    "\n",
    "\n",
    "class MolecularGraphNeuralNetwork(nn.Module):\n",
    "    def __init__(self, N_fingerprints, dim, layer_hidden, layer_output, nfeat, nhid, nclass, dropout):\n",
    "        super(MolecularGraphNeuralNetwork, self).__init__()\n",
    "        self.embed_fingerprint = nn.Embedding(N_fingerprints, dim)\n",
    "        self.W_fingerprint = nn.ModuleList([nn.Linear(dim, dim)\n",
    "                                            for _ in range(layer_hidden)])\n",
    "        self.W_output = nn.ModuleList([nn.Linear(dim+dim+dim+dim, dim+dim+dim+dim)\n",
    "                                       for _ in range(layer_output)])\n",
    "        if task == 'classification':\n",
    "            self.W_property = nn.Linear(dim, 2)\n",
    "        if task == 'regression':\n",
    "            self.W_property = nn.Linear(dim+dim+dim+dim, 1)\n",
    "        \n",
    "        # gcn\n",
    "        self.gc1 = GraphConvolution(nfeat, nhid)\n",
    "        self.gc2 = GraphConvolution(nhid, 128)\n",
    "        self.dropout = dropout\n",
    "        self.output = nn.Linear(128, 100).float()\n",
    "\n",
    "    def pad(self, matrices, pad_value):\n",
    "        \"\"\"Pad the list of matrices\n",
    "        with a pad_value (e.g., 0) for batch processing.\n",
    "        For example, given a list of matrices [A, B, C],\n",
    "        we obtain a new matrix [A00, 0B0, 00C],\n",
    "        where 0 is the zero (i.e., pad value) matrix.\n",
    "        \"\"\"\n",
    "        shapes = [m.shape for m in matrices]\n",
    "        M, N = sum([s[0] for s in shapes]), sum([s[1] for s in shapes])\n",
    "        zeros = torch.FloatTensor(np.zeros((M, N))).to(device)\n",
    "        pad_matrices = pad_value + zeros\n",
    "        i, j = 0, 0\n",
    "        for k, matrix in enumerate(matrices):\n",
    "            m, n = shapes[k]\n",
    "            pad_matrices[i:i+m, j:j+n] = matrix\n",
    "            i += m\n",
    "            j += n\n",
    "        return pad_matrices\n",
    "\n",
    "    def update(self, matrix, vectors, layer):\n",
    "        hidden_vectors = torch.relu(self.W_fingerprint[layer](vectors))\n",
    "        return hidden_vectors + torch.matmul(matrix, hidden_vectors)\n",
    "\n",
    "    def sum(self, vectors, axis):\n",
    "        sum_vectors = [torch.sum(v, 0) for v in torch.split(vectors, axis)]\n",
    "        return torch.stack(sum_vectors)\n",
    "\n",
    "    def mean(self, vectors, axis):\n",
    "        mean_vectors = [torch.mean(v, 0) for v in torch.split(vectors, axis)]\n",
    "        return torch.stack(mean_vectors)\n",
    "\n",
    "    def gnn(self, inputs):\n",
    "\n",
    "        \"\"\"Cat or pad each input data for batch processing.\"\"\"\n",
    "        fingerprints, adjacencies, molecular_sizes = inputs\n",
    "        fingerprints = torch.cat(fingerprints)\n",
    "        adjacencies = self.pad(adjacencies, 0)\n",
    "\n",
    "        \"\"\"GNN layer (update the fingerprint vectors).\"\"\"\n",
    "        fingerprint_vectors = self.embed_fingerprint(fingerprints)\n",
    "        for l in range(layer_hidden):\n",
    "            hs = self.update(adjacencies, fingerprint_vectors, l)\n",
    "            fingerprint_vectors = F.normalize(hs, 2, 1)  # normalize.\n",
    "\n",
    "        \"\"\"Molecular vector by sum or mean of the fingerprint vectors.\"\"\"\n",
    "        molecular_vectors = self.sum(fingerprint_vectors, molecular_sizes)\n",
    "        # molecular_vectors = self.mean(fingerprint_vectors, molecular_sizes)\n",
    "\n",
    "        return molecular_vectors\n",
    "\n",
    "    def mlp(self, vectors):\n",
    "        \"\"\"Classifier or regressor based on multilayer perceptron.\"\"\"\n",
    "        for l in range(layer_output):\n",
    "#             print(self.W_output[l](vectors))\n",
    "#             print(self.W_output[l](vectors).shape)\n",
    "            vectors = torch.relu(self.W_output[l](vectors))\n",
    "        outputs = self.W_property(vectors)\n",
    "        return outputs\n",
    "\n",
    "    def forward_classifier(self, data_batch, train):\n",
    "\n",
    "        inputs = data_batch[:-1]\n",
    "        correct_labels = torch.cat(data_batch[-1])\n",
    "\n",
    "        if train:\n",
    "            molecular_vectors = self.gnn(inputs)\n",
    "            predicted_scores = self.mlp(molecular_vectors)\n",
    "            loss = F.cross_entropy(predicted_scores, correct_labels)\n",
    "            return loss\n",
    "        else:\n",
    "            with torch.no_grad():\n",
    "                molecular_vectors = self.gnn(inputs)\n",
    "                predicted_scores = self.mlp(molecular_vectors)\n",
    "            predicted_scores = predicted_scores.to('cpu').data.numpy()\n",
    "            predicted_scores = [s[1] for s in predicted_scores]\n",
    "            correct_labels = correct_labels.to('cpu').data.numpy()\n",
    "            return predicted_scores, correct_labels\n",
    "\n",
    "    def forward_regressor(self, data_batch, x, adj, train):\n",
    "\n",
    "        inputs_d1 = data_batch[:3]\n",
    "        inputs_d2 = data_batch[3:-1]\n",
    "        correct_values = torch.cat(data_batch[-1])\n",
    "        \n",
    "        if train:\n",
    "            molecular_vectors_d1 = self.gnn(inputs_d1)\n",
    "            molecular_vectors_d2 = self.gnn(inputs_d2)\n",
    "#             molecular_vectors = np.concatenate([molecular_vectors_d1,molecular_vectors_d2])\n",
    "            # print(molecular_vectors_d1.shape, molecular_vectors_d1.shape)\n",
    "            x = F.relu(self.gc1(x, adj))\n",
    "            x = F.dropout(x, self.dropout, training=self.training)\n",
    "            x = self.gc2(x, adj)\n",
    "            x = self.output(x.float())\n",
    "            \n",
    "            molecular_vectors = torch.cat((molecular_vectors_d1, molecular_vectors_d2, x), 1)\n",
    "            # print(molecular_vectors.shape)\n",
    "            predicted_values = self.mlp(molecular_vectors)\n",
    "#             print(predicted_values.shape, correct_values.shape, molecular_vectors.shape, molecular_vectors_d1.shape)\n",
    "            # print(predicted_values.shape)\n",
    "            loss = F.mse_loss(predicted_values, correct_values)\n",
    "            return loss\n",
    "        # else:\n",
    "        #     with torch.no_grad():\n",
    "        #         molecular_vectors_d1 = self.gnn(inputs_d1)\n",
    "        #         molecular_vectors_d2 = self.gnn(inputs_d2)\n",
    "        #         molecular_vectors = torch.cat((molecular_vectors_d1, molecular_vectors_d2), 1)\n",
    "        #         predicted_values = self.mlp(molecular_vectors)\n",
    "        #     predicted_values = predicted_values.to('cpu').data.numpy()\n",
    "        #     correct_values = correct_values.to('cpu').data.numpy()\n",
    "        #     predicted_values = np.concatenate(predicted_values)\n",
    "        #     correct_values = np.concatenate(correct_values)\n",
    "        #     return predicted_values, correct_values\n",
    "\n",
    "\n",
    "class Trainer(object):\n",
    "    def __init__(self, model):\n",
    "        self.model = model\n",
    "        self.optimizer = optim.Adam(self.model.parameters(), lr=lr)\n",
    "\n",
    "    def train(self, dataset, x, adj, train = True):\n",
    "        np.random.shuffle(dataset)\n",
    "        N = len(dataset)\n",
    "        loss_total = 0\n",
    "        data_batch = list(zip(*dataset[:]))\n",
    "        if task == 'regression':\n",
    "          loss = self.model.forward_regressor(data_batch, x, adj, train=True)\n",
    "          self.optimizer.zero_grad()\n",
    "          loss.backward()\n",
    "          self.optimizer.step()\n",
    "          loss_total += loss.item()\n",
    "        # for i in range(0, N, batch_train):\n",
    "        #     data_batch = list(zip(*dataset[i:i+batch_train]))\n",
    "        #     if task == 'classification':\n",
    "        #         loss = self.model.forward_classifier(data_batch, train=True)\n",
    "        #     if task == 'regression':\n",
    "        #         loss = self.model.forward_regressor(data_batch, train=True)\n",
    "        #     self.optimizer.zero_grad()\n",
    "        #     loss.backward()\n",
    "        #     self.optimizer.step()\n",
    "        #     loss_total += loss.item()\n",
    "        return loss_total\n",
    "\n",
    "\n",
    "class Tester(object):\n",
    "    def __init__(self, model):\n",
    "        self.model = model\n",
    "\n",
    "    def test_regressor(self, dataset):\n",
    "        N = len(dataset)\n",
    "        SAE = 0  # sum absolute error.\n",
    "        for i in range(0, N, batch_test):\n",
    "            data_batch = list(zip(*dataset[i:i+batch_test]))\n",
    "            predicted_values, correct_values = self.model.forward_regressor(\n",
    "                                               data_batch, train=False)\n",
    "            SAE += sum(np.abs(predicted_values-correct_values))\n",
    "        MAE = SAE / N  # mean absolute error.\n",
    "        return MAE\n",
    "\n",
    "    def save_result(self, result, filename):\n",
    "        with open(filename, 'a') as f:\n",
    "            f.write(result + '\\n')\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "38KI95eFhzum"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "rZfBrByMhzxH"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "oQtu5NsRh3-X"
   },
   "outputs": [],
   "source": [
    "import math\n",
    "\n",
    "import torch\n",
    "\n",
    "from torch.nn.parameter import Parameter\n",
    "from torch.nn.modules.module import Module\n",
    "\n",
    "\n",
    "class GraphConvolution(Module):\n",
    "    \"\"\"\n",
    "    Simple GCN layer, similar to https://arxiv.org/abs/1609.02907\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, in_features, out_features, bias=True):\n",
    "        super(GraphConvolution, self).__init__()\n",
    "        self.in_features = in_features\n",
    "        self.out_features = out_features\n",
    "        self.weight = Parameter(torch.FloatTensor(in_features, out_features))\n",
    "        if bias:\n",
    "            self.bias = Parameter(torch.FloatTensor(out_features))\n",
    "        else:\n",
    "            self.register_parameter('bias', None)\n",
    "        self.reset_parameters()\n",
    "\n",
    "    def reset_parameters(self):\n",
    "        stdv = 1. / math.sqrt(self.weight.size(1))\n",
    "        self.weight.data.uniform_(-stdv, stdv)\n",
    "        if self.bias is not None:\n",
    "            self.bias.data.uniform_(-stdv, stdv)\n",
    "\n",
    "    def forward(self, input, adj):\n",
    "        support = torch.mm(input.double(), self.weight.double())\n",
    "        adj = torch.from_numpy(adj).double()\n",
    "#         print(type(adj), type(support))\n",
    "        # print(adj.shape, support.shape)\n",
    "        output = torch.spmm(adj, support)\n",
    "        if self.bias is not None:\n",
    "            return output + self.bias\n",
    "        else:\n",
    "            return output\n",
    "\n",
    "    def __repr__(self):\n",
    "        return self.__class__.__name__ + ' (' \\\n",
    "               + str(self.in_features) + ' -> ' \\\n",
    "               + str(self.out_features) + ')'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "0ypgO5h1h4Cf"
   },
   "outputs": [],
   "source": [
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "\n",
    "class GCN(nn.Module):\n",
    "    def __init__(self, nfeat, nhid, nclass, dropout):\n",
    "        super(GCN, self).__init__()\n",
    "\n",
    "        self.gc1 = GraphConvolution(nfeat, nhid)\n",
    "        self.gc2 = GraphConvolution(nhid, 128)\n",
    "        self.dropout = dropout\n",
    "        # self.output = nn.Linear(128, nclass).float()\n",
    "        self.output = nn.Linear(128, 100).float()\n",
    "\n",
    "    def forward(self, x, adj):\n",
    "#         x = x.astype('double')\n",
    "#         adj = adj.astype('double')\n",
    "        \n",
    "        x = F.relu(self.gc1(x, adj))\n",
    "        x = F.dropout(x, self.dropout, training=self.training)\n",
    "        x = self.gc2(x, adj)\n",
    "        x = self.output(x.float())\n",
    "#         print(x)\n",
    "        \n",
    "        return x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "8iG9NWqUjusu"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "MAuUvo1Kjuw2"
   },
   "outputs": [],
   "source": [
    "# features_train.shape\n",
    "# # Create dummy adjacency matrix for now\n",
    "# import scipy.sparse as sp\n",
    "# adj = sp.coo_matrix((len(data_for_gnn_gcn[:100]), len(data_for_gnn_gcn[:100])), dtype=np.float32).toarray()\n",
    "\n",
    "# # build symmetric adjacency matrix\n",
    "# adj = adj + np.matmul(adj.T, adj.T > adj) - np.matmul(adj, adj.T > adj)\n",
    "# adj\n",
    "# adj.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "PM454PP6U_Lh"
   },
   "outputs": [],
   "source": [
    "# # calculate adj\n",
    "# import scipy.sparse as sp\n",
    "# adj = sp.coo_matrix((100, 100), dtype=np.float32).toarray()\n",
    "# # build symmetric adjacency matrix\n",
    "# adj = adj + np.matmul(adj.T, adj.T > adj) - np.matmul(adj, adj.T > adj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "PfkavpKwSLkW"
   },
   "outputs": [],
   "source": [
    "def get_gcn_dataset(adj):\n",
    "  data_for_gcn = data_for_gnn_gcn['cell_line_expr']\n",
    "  gcn_train_data, gcn_dev_data = split_dataset(data_for_gcn, 0.9)\n",
    "  # gcn_train_data = gcn_train_data\n",
    "  # gcn_dev_data = gcn_dev_data\n",
    "  gcn_test_data = gcn_dev_data\n",
    "  # print(gcn_train_data.shape, gcn_dev_data.shape)\n",
    "\n",
    "  import scipy.sparse as sp\n",
    "  features_all = pd.DataFrame(gcn_train_data.to_list()).values\n",
    "  features_train = sp.csr_matrix(features_all, dtype=np.float32)\n",
    "  features_train = normalize(features_train)\n",
    "  features_train = torch.FloatTensor(np.array(features_train.todense()))\n",
    "\n",
    "  features_all = pd.DataFrame(gcn_dev_data.to_list()).values\n",
    "  features_dev = sp.csr_matrix(features_all, dtype=np.float32)\n",
    "  features_dev = normalize(features_dev)\n",
    "  features_dev = torch.FloatTensor(np.array(features_dev.todense()))\n",
    "\n",
    "  features_all = pd.DataFrame(gcn_test_data.to_list()).values\n",
    "  features_test = sp.csr_matrix(features_all, dtype=np.float32)\n",
    "  features_test = normalize(features_test)\n",
    "  features_test = torch.FloatTensor(np.array(features_test.todense()))\n",
    "\n",
    "  print('gcn input train, dev, test size: ', features_train.shape, features_dev.shape, features_test.shape )\n",
    "  \n",
    "  adj = normalize(adj + sp.eye(adj.shape[0]))\n",
    "\n",
    "  \n",
    "  labels = torch.LongTensor(data_for_gnn_gcn['synergy'])\n",
    "  adj1 = sparse_mx_to_torch_sparse_tensor(sp.csr_matrix(adj))\n",
    "\n",
    "  idx_train = gcn_train_data.index\n",
    "  idx_val = gcn_dev_data.index\n",
    "  idx_test = gcn_test_data.index\n",
    "\n",
    "  idx_train = torch.LongTensor(idx_train)\n",
    "  idx_val = torch.LongTensor(idx_val)\n",
    "  idx_test = torch.LongTensor(idx_test)\n",
    "\n",
    "  return features_train, features_dev, features_test, idx_train, idx_val, idx_test, adj\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "zstufQ43SLnW"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "i1Vlrby6SLqv"
   },
   "outputs": [],
   "source": [
    "\n",
    "# Run the main code function: \n",
    "task = 'regression'\n",
    "dataset = ''\n",
    "radius = 3\n",
    "dim = 50\n",
    "layer_hidden=6\n",
    "layer_output=6\n",
    "\n",
    "batch_train=32\n",
    "batch_test=32\n",
    "lr=1e-4\n",
    "lr_decay=0.99\n",
    "decay_interval=10\n",
    "iteration=5\n",
    "\n",
    "#     (task, dataset, radius, dim, layer_hidden, layer_output,\n",
    "#      batch_train, batch_test, lr, lr_decay, decay_interval, iteration,\n",
    "#      setting) = sys.argv[1:]\n",
    "(radius, dim, layer_hidden, layer_output,\n",
    "  batch_train, batch_test, decay_interval,\n",
    "  iteration) = map(int, [radius, dim, layer_hidden, layer_output,\n",
    "                        batch_train, batch_test,\n",
    "                        decay_interval, iteration])\n",
    "lr, lr_decay = map(float, [lr, lr_decay])\n",
    "#     device = torch.device('cuda')\n",
    "# device = torch.device('cpu')\n",
    "\n",
    "if torch.cuda.is_available():\n",
    "    device = torch.device('cuda')\n",
    "    print('The code uses a GPU!')\n",
    "else:\n",
    "    device = torch.device('cpu')\n",
    "    print('The code uses a CPU...')\n",
    "#     print('-'*100)\n",
    "\n",
    "print('Preprocessing the GNN', dataset, 'dataset.')\n",
    "print('Just a moment......')\n",
    "(dataset_train, dataset_dev, dataset_test,\n",
    "  N_fingerprints, final_index) = create_datasets(task, data_for_gnn_gcn, radius, device)\n",
    "print('-'*100)\n",
    "\n",
    "# dataset_train = dataset_train[:100]\n",
    "# dataset_dev = dataset_dev[-500:]\n",
    "\n",
    "print('The preprocess has finished!')\n",
    "print('# of training data samples:', len(dataset_train))\n",
    "print('# of development data samples:', len(dataset_dev))\n",
    "print('# of test data samples:', len(dataset_test))\n",
    "print('-'*100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "zXLPcPZRSScT"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "DlWhGyLASSgK"
   },
   "outputs": [],
   "source": [
    "# Remove the index which were skipped\n",
    "data_for_gnn_gcn = data_for_gnn_gcn.iloc[final_index].reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "bvsOScdwSXjC"
   },
   "outputs": [],
   "source": [
    "print('Preprocessing the GCN dataset.')\n",
    "print('Just a moment......')\n",
    "features_train, features_dev, features_test, idx_train, idx_val, idx_test, adj = get_gcn_dataset(adj)\n",
    "print('-'*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "-kh3HyJvSXnG"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Syghk-dXSXqN"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "anUPFoq0sJtQ",
    "outputId": "0b156e92-8a61-4e3c-aac2-11f7b91cf3d1"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating a model.\n",
      "# of model parameters: 519593\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Start training.\n",
      "The result is saved in the output directory every epoch!\n",
      "79\n",
      "0\n",
      "done\n",
      "0\t2.9097195320000537\t513.2335815429688\t0\t0\n",
      "1\n",
      "done\n",
      "The training will finish in about 0 hours 7 minutes.\n",
      "----------------------------------------------------------------------------------------------------\n",
      "0\t2.9097195320000537\t513.2335815429688\t0\t0\n",
      "1\t5.4772163469997395\t398.6176452636719\t0\t0\n",
      "2\n",
      "done\n",
      "2\t8.019814188000055\t321.8485107421875\t0\t0\n",
      "3\n",
      "done\n",
      "3\t10.437578408999798\t319.814697265625\t0\t0\n",
      "4\n",
      "done\n",
      "4\t12.863119435999579\t752.2893676757812\t0\t0\n",
      "5\n",
      "done\n",
      "5\t15.368189077999887\t563.160888671875\t0\t0\n",
      "6\n",
      "done\n",
      "6\t17.80202894499962\t701.803955078125\t0\t0\n",
      "7\n",
      "done\n",
      "7\t20.70601590899969\t451.8525085449219\t0\t0\n",
      "8\n",
      "done\n",
      "8\t23.1158027219999\t620.16455078125\t0\t0\n",
      "9\n",
      "done\n",
      "9\t25.285636264999994\t456.2766418457031\t0\t0\n",
      "10\n",
      "done\n",
      "10\t27.925222621000103\t853.5056762695312\t0\t0\n",
      "11\n",
      "done\n",
      "11\t30.817600622999635\t681.81005859375\t0\t0\n",
      "12\n",
      "done\n",
      "12\t33.10483297399969\t417.8182678222656\t0\t0\n",
      "13\n",
      "done\n",
      "13\t35.64969551299964\t444.107666015625\t0\t0\n",
      "14\n",
      "done\n",
      "14\t38.4685633199997\t398.5043640136719\t0\t0\n",
      "15\n",
      "done\n",
      "15\t40.870672445999844\t330.2064208984375\t0\t0\n",
      "16\n",
      "done\n",
      "16\t43.152801338000245\t607.6362915039062\t0\t0\n",
      "17\n",
      "done\n",
      "17\t45.3531317879997\t663.0324096679688\t0\t0\n",
      "18\n",
      "done\n",
      "18\t47.93499896599951\t789.7003173828125\t0\t0\n",
      "19\n",
      "done\n",
      "19\t50.512685613999565\t513.624267578125\t0\t0\n",
      "20\n",
      "done\n",
      "20\t52.97699535499942\t468.3809509277344\t0\t0\n",
      "21\n",
      "done\n",
      "21\t55.609589941000195\t629.532958984375\t0\t0\n",
      "22\n",
      "done\n",
      "22\t57.89745201000005\t591.2472534179688\t0\t0\n",
      "23\n",
      "done\n",
      "23\t60.34847077699942\t282.4077453613281\t0\t0\n",
      "24\n",
      "done\n",
      "24\t63.03662079500009\t439.6694030761719\t0\t0\n",
      "25\n",
      "done\n",
      "25\t65.51353663200007\t788.9495239257812\t0\t0\n",
      "26\n",
      "done\n",
      "26\t67.97239788200022\t752.338134765625\t0\t0\n",
      "27\n",
      "done\n",
      "27\t69.97596868599976\t891.5079956054688\t0\t0\n",
      "28\n",
      "done\n",
      "28\t72.44458471500002\t654.5595092773438\t0\t0\n",
      "29\n",
      "done\n",
      "29\t74.87852355199993\t767.90625\t0\t0\n",
      "30\n",
      "done\n",
      "30\t77.39096847200017\t683.936767578125\t0\t0\n",
      "31\n",
      "done\n",
      "31\t79.5865050020002\t373.3685302734375\t0\t0\n",
      "32\n",
      "done\n",
      "32\t82.10844263800027\t554.7952270507812\t0\t0\n",
      "33\n",
      "done\n",
      "33\t84.79283835200022\t901.1317138671875\t0\t0\n",
      "34\n",
      "done\n",
      "34\t87.32765759799986\t526.0061645507812\t0\t0\n",
      "35\n",
      "done\n",
      "35\t89.3549158400001\t252.93789672851562\t0\t0\n",
      "36\n",
      "done\n",
      "36\t91.81852616600008\t777.0966186523438\t0\t0\n",
      "37\n",
      "done\n",
      "37\t94.06338059100017\t832.1743774414062\t0\t0\n",
      "38\n",
      "done\n",
      "38\t96.33117498399952\t389.5570373535156\t0\t0\n",
      "39\n",
      "done\n",
      "39\t98.95664696999938\t439.1918029785156\t0\t0\n",
      "40\n",
      "done\n",
      "40\t101.19376669299982\t325.4663391113281\t0\t0\n",
      "41\n",
      "done\n",
      "41\t103.76731070200003\t661.572998046875\t0\t0\n",
      "42\n",
      "done\n",
      "42\t106.18263597299938\t307.0623779296875\t0\t0\n",
      "43\n",
      "done\n",
      "43\t108.464661084\t471.8548583984375\t0\t0\n",
      "44\n",
      "done\n",
      "44\t110.75768145500024\t390.3077697753906\t0\t0\n",
      "45\n",
      "done\n",
      "45\t113.16643309600022\t479.3379821777344\t0\t0\n",
      "46\n",
      "done\n",
      "46\t115.48090907099959\t334.86480712890625\t0\t0\n",
      "47\n",
      "done\n",
      "47\t117.99282900700018\t389.501220703125\t0\t0\n",
      "48\n",
      "done\n",
      "48\t120.18854685199949\t278.5444641113281\t0\t0\n",
      "49\n",
      "done\n",
      "49\t122.45024628099964\t473.8418884277344\t0\t0\n",
      "50\n",
      "done\n",
      "50\t125.18884544100001\t408.4710693359375\t0\t0\n",
      "51\n",
      "done\n",
      "51\t127.54890595399957\t262.771240234375\t0\t0\n",
      "52\n",
      "done\n",
      "52\t129.79318555899954\t494.8360290527344\t0\t0\n",
      "53\n",
      "done\n",
      "53\t132.09069083199938\t967.0516357421875\t0\t0\n",
      "54\n",
      "done\n",
      "54\t134.41285313000026\t670.142822265625\t0\t0\n",
      "55\n",
      "done\n",
      "55\t136.764634268\t338.3493347167969\t0\t0\n",
      "56\n",
      "done\n",
      "56\t139.47641702999954\t467.6245422363281\t0\t0\n",
      "57\n",
      "done\n",
      "57\t141.5803108379996\t366.1156311035156\t0\t0\n",
      "58\n",
      "done\n",
      "58\t144.2891137329998\t792.4996337890625\t0\t0\n",
      "59\n",
      "done\n",
      "59\t146.99169669399998\t609.684326171875\t0\t0\n",
      "60\n",
      "done\n",
      "60\t149.47113825899942\t349.1280822753906\t0\t0\n",
      "61\n",
      "done\n",
      "61\t152.11383839500013\t399.9644470214844\t0\t0\n",
      "62\n",
      "done\n",
      "62\t154.60647856499963\t331.415283203125\t0\t0\n",
      "63\n",
      "done\n",
      "63\t156.88922719899983\t362.68328857421875\t0\t0\n",
      "64\n",
      "done\n",
      "64\t159.04540951199942\t576.8154907226562\t0\t0\n",
      "65\n",
      "done\n",
      "65\t161.5595311039997\t371.7618408203125\t0\t0\n",
      "66\n",
      "done\n",
      "66\t164.0002949869995\t502.5982360839844\t0\t0\n",
      "67\n",
      "done\n",
      "67\t166.43001621799976\t245.63360595703125\t0\t0\n",
      "68\n",
      "done\n",
      "68\t169.04799006899975\t428.4706726074219\t0\t0\n",
      "69\n",
      "done\n",
      "69\t171.3311329090002\t289.2734680175781\t0\t0\n",
      "70\n",
      "done\n",
      "70\t173.6717827299999\t397.294921875\t0\t0\n",
      "71\n",
      "done\n",
      "71\t175.86456626200015\t315.03936767578125\t0\t0\n",
      "72\n",
      "done\n",
      "72\t178.21945285099991\t351.59490966796875\t0\t0\n",
      "73\n",
      "done\n",
      "73\t180.613119138\t502.7274169921875\t0\t0\n",
      "74\n",
      "done\n",
      "74\t183.18783603799966\t528.9769287109375\t0\t0\n",
      "75\n",
      "done\n",
      "75\t185.34891926099954\t329.3954162597656\t0\t0\n",
      "76\n",
      "done\n",
      "76\t187.5555693259994\t623.9275512695312\t0\t0\n",
      "77\n",
      "done\n",
      "77\t190.00184295100007\t441.09820556640625\t0\t0\n",
      "78\n",
      "done\n",
      "78\t192.03474993100008\t408.8433532714844\t0\t0\n",
      "79\n",
      "done\n",
      "79\t194.21342516599998\t541.3194580078125\t0\t0\n",
      "80\n",
      "done\n",
      "80\t196.63449676500022\t386.2417907714844\t0\t0\n",
      "81\n",
      "done\n",
      "81\t199.08744179899986\t254.26646423339844\t0\t0\n",
      "82\n",
      "done\n",
      "82\t201.6957986440002\t304.14276123046875\t0\t0\n",
      "83\n",
      "done\n",
      "83\t204.32207407299938\t656.1868896484375\t0\t0\n",
      "84\n",
      "done\n",
      "84\t206.72675264600002\t515.9363403320312\t0\t0\n",
      "85\n",
      "done\n",
      "85\t209.13762995300021\t640.8056030273438\t0\t0\n",
      "86\n",
      "done\n",
      "86\t211.78627997999956\t372.0946960449219\t0\t0\n",
      "87\n",
      "done\n",
      "87\t214.1390201539998\t588.6446533203125\t0\t0\n",
      "88\n",
      "done\n",
      "88\t216.43146735000028\t381.6555480957031\t0\t0\n",
      "89\n",
      "done\n",
      "89\t218.7111551150001\t749.4279174804688\t0\t0\n",
      "90\n",
      "done\n",
      "90\t221.37256052700013\t607.8008422851562\t0\t0\n",
      "91\n",
      "done\n",
      "91\t223.6198381969998\t370.02203369140625\t0\t0\n",
      "92\n",
      "done\n",
      "92\t226.13647669300008\t379.3359375\t0\t0\n",
      "93\n",
      "done\n",
      "93\t228.5504462580002\t317.860595703125\t0\t0\n",
      "94\n",
      "done\n",
      "94\t230.70063996099998\t268.2979431152344\t0\t0\n",
      "95\n",
      "done\n",
      "95\t232.97303751599975\t515.3425903320312\t0\t0\n",
      "96\n",
      "done\n",
      "96\t235.36749543899987\t588.4891967773438\t0\t0\n",
      "97\n",
      "done\n",
      "97\t238.03277922599955\t811.457763671875\t0\t0\n",
      "98\n",
      "done\n",
      "98\t240.6694381520001\t452.4961853027344\t0\t0\n",
      "99\n",
      "done\n",
      "99\t243.38880324399997\t368.1745300292969\t0\t0\n",
      "100\n",
      "done\n",
      "100\t245.8849830979998\t532.6228637695312\t0\t0\n",
      "101\n",
      "done\n",
      "101\t248.0965422549998\t493.7364196777344\t0\t0\n",
      "102\n",
      "done\n",
      "102\t250.4438287639996\t211.38006591796875\t0\t0\n",
      "103\n",
      "done\n",
      "103\t252.9407579990002\t324.7474670410156\t0\t0\n",
      "104\n",
      "done\n",
      "104\t255.2263444780001\t741.3677368164062\t0\t0\n",
      "105\n",
      "done\n",
      "105\t257.6843150099994\t614.9178466796875\t0\t0\n",
      "106\n",
      "done\n",
      "106\t259.7947942519995\t906.1565551757812\t0\t0\n",
      "107\n",
      "done\n",
      "107\t262.3990572929997\t575.70556640625\t0\t0\n",
      "108\n",
      "done\n",
      "108\t264.55142113500006\t753.1788330078125\t0\t0\n",
      "109\n",
      "done\n",
      "109\t267.0937848579997\t665.2984619140625\t0\t0\n",
      "110\n",
      "done\n",
      "110\t269.5835826679995\t336.903076171875\t0\t0\n",
      "111\n",
      "done\n",
      "111\t271.91185146599946\t503.9646911621094\t0\t0\n",
      "112\n",
      "done\n",
      "112\t274.0445644219999\t847.5309448242188\t0\t0\n",
      "113\n",
      "done\n",
      "113\t276.3976341079997\t502.9288330078125\t0\t0\n",
      "114\n",
      "done\n",
      "114\t278.62983149999945\t221.9243621826172\t0\t0\n",
      "115\n",
      "done\n",
      "115\t281.0613368539998\t753.1436767578125\t0\t0\n",
      "116\n",
      "done\n",
      "116\t283.4373170509998\t822.5357666015625\t0\t0\n",
      "117\n",
      "done\n",
      "117\t285.8601413460001\t345.1571044921875\t0\t0\n",
      "118\n",
      "done\n",
      "118\t288.35493322899947\t401.2335205078125\t0\t0\n",
      "119\n",
      "done\n",
      "119\t290.95061133299987\t299.9876403808594\t0\t0\n",
      "120\n",
      "done\n",
      "120\t293.62613765399965\t590.641357421875\t0\t0\n",
      "121\n",
      "done\n",
      "121\t295.96039360099985\t281.7543640136719\t0\t0\n",
      "122\n",
      "done\n",
      "122\t298.09119221199944\t491.0826110839844\t0\t0\n",
      "123\n",
      "done\n",
      "123\t300.5897011699999\t380.6239013671875\t0\t0\n",
      "124\n",
      "done\n",
      "124\t303.06567143600023\t447.94866943359375\t0\t0\n",
      "125\n",
      "done\n",
      "125\t305.4594236639996\t327.7395324707031\t0\t0\n",
      "126\n",
      "done\n",
      "126\t307.74182898399977\t375.5101623535156\t0\t0\n",
      "127\n",
      "done\n",
      "127\t309.888995368\t272.37127685546875\t0\t0\n",
      "128\n",
      "done\n",
      "128\t312.1014175720002\t463.3389892578125\t0\t0\n",
      "129\n",
      "done\n",
      "129\t314.7888606159995\t411.938232421875\t0\t0\n",
      "130\n",
      "done\n",
      "130\t317.0162636139994\t258.3066711425781\t0\t0\n",
      "131\n",
      "done\n",
      "131\t319.27856093599985\t493.2989501953125\t0\t0\n",
      "132\n",
      "done\n",
      "132\t321.67355934199986\t963.48876953125\t0\t0\n",
      "133\n",
      "done\n",
      "133\t323.83832140100003\t666.3585205078125\t0\t0\n",
      "134\n",
      "done\n",
      "134\t326.17811729499954\t334.63226318359375\t0\t0\n",
      "135\n",
      "done\n",
      "135\t328.7749428389998\t463.84429931640625\t0\t0\n",
      "136\n",
      "done\n",
      "136\t331.0506422460003\t356.4590148925781\t0\t0\n",
      "137\n",
      "done\n",
      "137\t333.68722765999973\t783.622314453125\t0\t0\n",
      "138\n",
      "done\n",
      "138\t336.1804865570002\t599.5769653320312\t0\t0\n",
      "139\n",
      "done\n",
      "139\t338.6178309299994\t350.119140625\t0\t0\n",
      "140\n",
      "done\n",
      "140\t341.0309985369995\t397.4591064453125\t0\t0\n",
      "141\n",
      "done\n",
      "141\t343.3629339979998\t331.5058288574219\t0\t0\n",
      "142\n",
      "done\n",
      "142\t345.71244593400024\t351.95294189453125\t0\t0\n",
      "143\n",
      "done\n",
      "143\t347.8734925919998\t572.5133666992188\t0\t0\n",
      "144\n",
      "done\n",
      "144\t350.09675270800017\t363.40313720703125\t0\t0\n",
      "145\n",
      "done\n",
      "145\t352.6922015529999\t497.6894836425781\t0\t0\n",
      "146\n",
      "done\n",
      "146\t355.1279469459996\t245.31338500976562\t0\t0\n",
      "147\n",
      "done\n",
      "147\t357.59110550100013\t424.1948547363281\t0\t0\n",
      "148\n",
      "done\n",
      "148\t360.05783272499957\t284.6788024902344\t0\t0\n",
      "149\n",
      "done\n",
      "149\t362.18690753699957\t391.6500244140625\t0\t0\n",
      "150\n",
      "done\n",
      "150\t364.34654271399995\t303.66326904296875\t0\t0\n",
      "151\n",
      "done\n",
      "151\t366.6700854969995\t343.8661804199219\t0\t0\n",
      "152\n",
      "done\n",
      "152\t368.9648511489995\t495.7193603515625\t0\t0\n",
      "153\n",
      "done\n",
      "153\t371.45503924199966\t515.034912109375\t0\t0\n",
      "154\n",
      "done\n",
      "154\t373.62303185899964\t320.2290954589844\t0\t0\n",
      "155\n",
      "done\n",
      "155\t375.72397374999946\t614.1900634765625\t0\t0\n",
      "156\n",
      "done\n",
      "156\t378.32156270199994\t439.7652282714844\t0\t0\n",
      "157\n",
      "done\n",
      "157\t380.349040737\t392.19598388671875\t0\t0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "################################################################## RUN MODELLL\n",
    "\n",
    "dim = 50\n",
    "iteration=50\n",
    "print('Creating a model.')\n",
    "torch.manual_seed(1234)\n",
    "\n",
    "\n",
    "model = MolecularGraphNeuralNetwork(\n",
    "        N_fingerprints, dim, layer_hidden, layer_output, nfeat=features_train.shape[1],nhid=16,nclass=1,dropout=0.2 ).to(device)\n",
    "trainer = Trainer(model)\n",
    "# tester = Tester(model)\n",
    "print('# of model parameters:',\n",
    "      sum([np.prod(p.size()) for p in model.parameters()]))\n",
    "print('-'*100)\n",
    "\n",
    "file_result = train_data_path + '/setting1_run.txt'\n",
    "\n",
    "if task == 'regression':\n",
    "    result = 'Epoch\\tTime(sec)\\tLoss_train\\tMAE_dev\\tMAE_test'\n",
    "\n",
    "with open(file_result, 'w') as f:\n",
    "    f.write(result + '\\n')\n",
    "\n",
    "print('Start training.')\n",
    "print('The result is saved in the output directory every epoch!')\n",
    "\n",
    "all_loss_train = []\n",
    "\n",
    "\n",
    "# np.random.seed(1234)\n",
    "\n",
    "start = timeit.default_timer()\n",
    "\n",
    "# BATCH THIS HERE AND ADJ MATRIX CALC EVERY TIME\n",
    "\n",
    "# epoch_batch_size = int( len(dataset_train)/iteration )\n",
    "epoch_batch_size = 100\n",
    "batch_start_nd = 0\n",
    "batch_end_nd = epoch_batch_size\n",
    "iteration = int(len(dataset_train)/epoch_batch_size)\n",
    "print(iteration)\n",
    "\n",
    "for epoch in range(iteration*2):\n",
    "\n",
    "    # Dont decay lr\n",
    "    # if epoch % decay_interval == 0:\n",
    "    #     trainer.optimizer.param_groups[0]['lr'] *= lr_decay\n",
    "\n",
    "    ds_train = dataset_train[batch_start_nd:batch_end_nd]\n",
    "    fs_train = features_train[batch_start_nd:batch_end_nd]\n",
    "\n",
    "\n",
    "    print(epoch)\n",
    "    loss_train = trainer.train(ds_train, fs_train, adj, train=True)\n",
    "    print('done')\n",
    "    # if task == 'regression':\n",
    "    #     prediction_dev = tester.test_regressor(dataset_dev)\n",
    "    #     prediction_test = tester.test_regressor(dataset_test)\n",
    "    prediction_test,prediction_dev = 0, 0\n",
    "\n",
    "    time = timeit.default_timer() - start\n",
    "\n",
    "    if epoch == 1:\n",
    "        minutes = time * iteration / 60\n",
    "        hours = int(minutes / 60)\n",
    "        minutes = int(minutes - 60 * hours)\n",
    "        print('The training will finish in about',\n",
    "              hours, 'hours', minutes, 'minutes.')\n",
    "        print('-'*100)\n",
    "        print(result)\n",
    "\n",
    "    result = '\\t'.join(map(str, [epoch, time, loss_train,\n",
    "                                 prediction_dev, prediction_test]))\n",
    "    # tester.save_result(result, file_result)\n",
    "\n",
    "    print(result)\n",
    "    all_loss_train.append(loss_train)\n",
    "    epoch += 1\n",
    "\n",
    "    batch_start_nd = batch_end_nd\n",
    "    batch_end_nd = batch_start_nd + epoch_batch_size\n",
    "    if(epoch == iteration):\n",
    "      batch_start_nd = 0\n",
    "      batch_end_nd = epoch_batch_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Rcscv5ezjwNN"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "u_CTvKzKjwQ8"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "jLd8ZKpJvH2g"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "o8Q-aJO6dDIS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "18ncRUCWIa2Z"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OgY309pNAjGg",
    "outputId": "7ca0b156-e87c-444f-b201-cce5fb7978c4"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The code uses a CPU...\n",
      "Preprocessing the GNN  dataset.\n",
      "Just a moment......\n",
      "(10000, 5)\n",
      "Skipped 1131 smiles in training data\n",
      "----------------------------------------------------------------------------------------------------\n",
      "The preprocess has finished!\n",
      "# of training data samples: 7982\n",
      "# of development data samples: 887\n",
      "# of test data samples: 887\n",
      "----------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "HT8ArtdhCwSk"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "8uhxdOiuCwhh"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "Rt4Dt4M9o3m_",
    "outputId": "b1e0082e-ea79-440c-fc2e-73f2c1c0d401"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preprocessing the GCN dataset.\n",
      "Just a moment......\n",
      "gcn input train, dev, test size:  torch.Size([7982, 10000]) torch.Size([887, 10000]) torch.Size([887, 10000])\n",
      "----------------------------------------------------------------------------------------------------\n",
      "gcn input train, dev, test size:  torch.Size([7982, 10000]) torch.Size([887, 10000]) torch.Size([887, 10000])\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "LLnPj9TkxVuq"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "TZaZlUeUxVxB",
    "outputId": "630c8ef0-2661-45df-c8b6-7c8456f1066d"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4500"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n = int(0.9 * 5000)\n",
    "n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "nYajInXCcEZf"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "frhJap4Jw_ww"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "epCdF4qfcEbn"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "ObyySR3GcEdr"
   },
   "outputs": [],
   "source": [
    "\n",
    "################################################################## RUN MODELLL\n",
    "\n",
    "dim = 50\n",
    "iteration=100\n",
    "print('Creating a model.')\n",
    "torch.manual_seed(1234)\n",
    "model = MolecularGraphNeuralNetwork(\n",
    "        N_fingerprints, dim, layer_hidden, layer_output, ).to(device)\n",
    "trainer = Trainer(model)\n",
    "tester = Tester(model)\n",
    "print('# of model parameters:',\n",
    "      sum([np.prod(p.size()) for p in model.parameters()]))\n",
    "print('-'*100)\n",
    "\n",
    "file_result = train_data_path + '/setting1_run.txt'\n",
    "\n",
    "if task == 'regression':\n",
    "    result = 'Epoch\\tTime(sec)\\tLoss_train\\tMAE_dev\\tMAE_test'\n",
    "\n",
    "with open(file_result, 'w') as f:\n",
    "    f.write(result + '\\n')\n",
    "\n",
    "print('Start training.')\n",
    "print('The result is saved in the output directory every epoch!')\n",
    "\n",
    "all_loss_train = []\n",
    "\n",
    "\n",
    "# np.random.seed(1234)\n",
    "\n",
    "start = timeit.default_timer()\n",
    "\n",
    "for epoch in range(iteration):\n",
    "\n",
    "    # Dont decay lr\n",
    "    # if epoch % decay_interval == 0:\n",
    "    #     trainer.optimizer.param_groups[0]['lr'] *= lr_decay\n",
    "\n",
    "    loss_train = trainer.train(dataset_train)\n",
    "\n",
    "    if task == 'regression':\n",
    "        prediction_dev = tester.test_regressor(dataset_dev)\n",
    "        prediction_test = tester.test_regressor(dataset_test)\n",
    "\n",
    "    time = timeit.default_timer() - start\n",
    "\n",
    "    if epoch == 1:\n",
    "        minutes = time * iteration / 60\n",
    "        hours = int(minutes / 60)\n",
    "        minutes = int(minutes - 60 * hours)\n",
    "        print('The training will finish in about',\n",
    "              hours, 'hours', minutes, 'minutes.')\n",
    "        print('-'*100)\n",
    "        print(result)\n",
    "\n",
    "    result = '\\t'.join(map(str, [epoch, time, loss_train,\n",
    "                                 prediction_dev, prediction_test]))\n",
    "    tester.save_result(result, file_result)\n",
    "\n",
    "    print(result)\n",
    "    all_loss_train.append(loss_train)\n",
    "    epoch += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QvToUDTOcEfp"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Oy1ydxUQcEg7"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "1_zQ4mwKcEi3"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "69FxrnB0cElO"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Rey_pIRicEnJ"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "_h6-29blcEpT"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "MlgH8Ml5cEqs"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "sl_nFh1ncEsh"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "e0-2InUCcEuV"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "koesXfC7cEwv"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "ly2Cej6VcEzo"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "F8F7VCZW4u0i"
   },
   "outputs": [],
   "source": [
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "xleiY0HNGIuR"
   },
   "outputs": [],
   "source": [
    "import torch\n",
    "torch.cuda.is_available()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "242SHvyRGPhA"
   },
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "tf.test.gpu_device_name()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "ySVkmnEIGRy0"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [],
   "name": "Drug Synergy Prediction",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
